{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "015e69ff",
   "metadata": {},
   "source": [
    "# Run Scrublet script (python)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb561d09",
   "metadata": {},
   "outputs": [],
   "source": [
    "# RNA doublet removal"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d8b5cc89",
   "metadata": {},
   "outputs": [],
   "source": [
    "# example\n",
    "# python3 scrublet.py -c /path/to/soupx/adjusted/matrices -o /path/to/scrublet/outdirectory/to/save/results\n",
    "# python3 scrublet.py -c /volume/carehf/snRNA/out/QC_preprocessing/JB352_soupX/ -o /volume/carehf/snRNA/out/QC_preprocessing/JB352_scrublet/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "513297ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "out_dir <- ''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c978de30",
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered_list <- list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "cd0e3eff",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"python3 scrublet.py -c /volume/afib/out/QC_preprocessing/AF_1_LA_soupX/ -o /volume/afib/out/QC_preprocessing/AF_1_LA_scrublet/\"\n",
      "[1] \"python3 scrublet.py -c /volume/afib/out/QC_preprocessing/AF_2_LA_soupX/ -o /volume/afib/out/QC_preprocessing/AF_2_LA_scrublet/\"\n",
      "[1] \"python3 scrublet.py -c /volume/afib/out/QC_preprocessing/AF_3_LA_soupX/ -o /volume/afib/out/QC_preprocessing/AF_3_LA_scrublet/\"\n",
      "[1] \"python3 scrublet.py -c /volume/afib/out/QC_preprocessing/AF_4_LA_soupX/ -o /volume/afib/out/QC_preprocessing/AF_4_LA_scrublet/\"\n",
      "[1] \"python3 scrublet.py -c /volume/afib/out/QC_preprocessing/AF_5_LA_soupX/ -o /volume/afib/out/QC_preprocessing/AF_5_LA_scrublet/\"\n",
      "[1] \"python3 scrublet.py -c /volume/afib/out/QC_preprocessing/AF_6_LA_soupX/ -o /volume/afib/out/QC_preprocessing/AF_6_LA_scrublet/\"\n",
      "[1] \"python3 scrublet.py -c /volume/afib/out/QC_preprocessing/AF_7_LA_soupX/ -o /volume/afib/out/QC_preprocessing/AF_7_LA_scrublet/\"\n",
      "[1] \"python3 scrublet.py -c /volume/afib/out/QC_preprocessing/AF_8_LA_soupX/ -o /volume/afib/out/QC_preprocessing/AF_8_LA_scrublet/\"\n",
      "[1] \"python3 scrublet.py -c /volume/afib/out/QC_preprocessing/CTRL_1_LA_soupX/ -o /volume/afib/out/QC_preprocessing/CTRL_1_LA_scrublet/\"\n",
      "[1] \"python3 scrublet.py -c /volume/afib/out/QC_preprocessing/CTRL_2_LA_soupX/ -o /volume/afib/out/QC_preprocessing/CTRL_2_LA_scrublet/\"\n",
      "[1] \"python3 scrublet.py -c /volume/afib/out/QC_preprocessing/CTRL_3_LA_soupX/ -o /volume/afib/out/QC_preprocessing/CTRL_3_LA_scrublet/\"\n",
      "[1] \"python3 scrublet.py -c /volume/afib/out/QC_preprocessing/CTRL_4_LA_soupX/ -o /volume/afib/out/QC_preprocessing/CTRL_4_LA_scrublet/\"\n",
      "[1] \"python3 scrublet.py -c /volume/afib/out/QC_preprocessing/CTRL_5_LA_soupX/ -o /volume/afib/out/QC_preprocessing/CTRL_5_LA_scrublet/\"\n",
      "[1] \"python3 scrublet.py -c /volume/afib/out/QC_preprocessing/CTRL_6_LA_soupX/ -o /volume/afib/out/QC_preprocessing/CTRL_6_LA_scrublet/\"\n",
      "[1] \"python3 scrublet.py -c /volume/afib/out/QC_preprocessing/CTRL_7_LA_soupX/ -o /volume/afib/out/QC_preprocessing/CTRL_7_LA_scrublet/\"\n",
      "[1] \"python3 scrublet.py -c /volume/afib/out/QC_preprocessing/CTRL_8_LA_soupX/ -o /volume/afib/out/QC_preprocessing/CTRL_8_LA_scrublet/\"\n",
      "[1] \"python3 scrublet.py -c /volume/afib/out/QC_preprocessing/D9_LA_soupX/ -o /volume/afib/out/QC_preprocessing/D9_LA_scrublet/\"\n",
      "[1] \"python3 scrublet.py -c /volume/afib/out/QC_preprocessing/D1_LA_soupX/ -o /volume/afib/out/QC_preprocessing/D1_LA_scrublet/\"\n",
      "[1] \"python3 scrublet.py -c /volume/afib/out/QC_preprocessing/HF6_LA_soupX/ -o /volume/afib/out/QC_preprocessing/HF6_LA_scrublet/\"\n",
      "[1] \"python3 scrublet.py -c /volume/afib/out/QC_preprocessing/HF9_LA_soupX/ -o /volume/afib/out/QC_preprocessing/HF9_LA_scrublet/\"\n"
     ]
    }
   ],
   "source": [
    "# create linux folders for the scrublet outputs if the folder doesn't already exist and write out the scrublet command line\n",
    "for (i in seq_along(filtered_list)){\n",
    "    sampleName <- names(filtered_list[i])\n",
    "    \n",
    "    folder <- paste0(out_dir,sampleName,\"_scrublet\")\n",
    "    \n",
    "    if (!(dir.exists(folder))) {\n",
    "        dir.create(folder)\n",
    "    }\n",
    "    \n",
    "    print(paste0('python3 scrublet.py -c ',out_dir,sampleName,\"_soupX/ -o \",folder,'/'))\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8095a3f4",
   "metadata": {},
   "source": [
    "# Run AMULET doublet removal script (bash)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a5e0e1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ATAC doublet removal"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fdb5e75a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# example:\n",
    "# AMULET.sh /path/to/fragments.tsv.gz /path/to/singlecell.csv /path/to/human_autosomes.txt /path/to/repeatfilter.bed /path/to/output/ /path/to/shellscript/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e479467",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Multiome cellranger output doesn't have singlecell.csv input file. Instead it has per_barcode_metrics.csv\n",
    "# Need to change the file and changename barcode to is__cell_barcode: \n",
    "# In linux terminal:\n",
    "# cat per_barcode_metrics.csv | cut -d, -f1,4 | sed \"1s/.*/barcode,is__cell_barcode/\" > singlecell.csv\n",
    "\n",
    "# or in R:\n",
    "# for(f in files){\n",
    "#    message(f)\n",
    "#    bc_metrics = read.csv(file.path(\"path\", \"to\", f, \"per_barcode_metrics.csv\")) #read in the per_barcode_metrics.csv\n",
    "#    df = data.frame(\"barcode\" = bc_metrics$barcode, \"is__cell_barcode\" =  bc_metrics$is_cell) #make a dataframe that can be used as \"single_cell.csv\"\n",
    "#    write.csv(df, file.path(\"path\", \"to\", f, \"single_cell.csv\"), row.names=FALSE, quote=FALSE) #write out the dataframe as \"single_cell.csv\"\n",
    "#    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "089a93a1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'cat AF_1_LA_per_barcode_metrics.csv | cut -d, -f1,4 | sed \"1s/.*/barcode,is__cell_barcode/\" &gt; AF_1_LA_singlecell.csv'</li><li>'cat AF_2_LA_per_barcode_metrics.csv | cut -d, -f1,4 | sed \"1s/.*/barcode,is__cell_barcode/\" &gt; AF_2_LA_singlecell.csv'</li><li>'cat AF_3_LA_per_barcode_metrics.csv | cut -d, -f1,4 | sed \"1s/.*/barcode,is__cell_barcode/\" &gt; AF_3_LA_singlecell.csv'</li><li>'cat AF_4_LA_per_barcode_metrics.csv | cut -d, -f1,4 | sed \"1s/.*/barcode,is__cell_barcode/\" &gt; AF_4_LA_singlecell.csv'</li><li>'cat AF_5_LA_per_barcode_metrics.csv | cut -d, -f1,4 | sed \"1s/.*/barcode,is__cell_barcode/\" &gt; AF_5_LA_singlecell.csv'</li><li>'cat AF_6_LA_per_barcode_metrics.csv | cut -d, -f1,4 | sed \"1s/.*/barcode,is__cell_barcode/\" &gt; AF_6_LA_singlecell.csv'</li><li>'cat AF_7_LA_per_barcode_metrics.csv | cut -d, -f1,4 | sed \"1s/.*/barcode,is__cell_barcode/\" &gt; AF_7_LA_singlecell.csv'</li><li>'cat AF_8_LA_per_barcode_metrics.csv | cut -d, -f1,4 | sed \"1s/.*/barcode,is__cell_barcode/\" &gt; AF_8_LA_singlecell.csv'</li><li>'cat CTRL_1_LA_per_barcode_metrics.csv | cut -d, -f1,4 | sed \"1s/.*/barcode,is__cell_barcode/\" &gt; CTRL_1_LA_singlecell.csv'</li><li>'cat CTRL_2_LA_per_barcode_metrics.csv | cut -d, -f1,4 | sed \"1s/.*/barcode,is__cell_barcode/\" &gt; CTRL_2_LA_singlecell.csv'</li><li>'cat CTRL_3_LA_per_barcode_metrics.csv | cut -d, -f1,4 | sed \"1s/.*/barcode,is__cell_barcode/\" &gt; CTRL_3_LA_singlecell.csv'</li><li>'cat CTRL_4_LA_per_barcode_metrics.csv | cut -d, -f1,4 | sed \"1s/.*/barcode,is__cell_barcode/\" &gt; CTRL_4_LA_singlecell.csv'</li><li>'cat CTRL_5_LA_per_barcode_metrics.csv | cut -d, -f1,4 | sed \"1s/.*/barcode,is__cell_barcode/\" &gt; CTRL_5_LA_singlecell.csv'</li><li>'cat CTRL_6_LA_per_barcode_metrics.csv | cut -d, -f1,4 | sed \"1s/.*/barcode,is__cell_barcode/\" &gt; CTRL_6_LA_singlecell.csv'</li><li>'cat CTRL_7_LA_per_barcode_metrics.csv | cut -d, -f1,4 | sed \"1s/.*/barcode,is__cell_barcode/\" &gt; CTRL_7_LA_singlecell.csv'</li><li>'cat CTRL_8_LA_per_barcode_metrics.csv | cut -d, -f1,4 | sed \"1s/.*/barcode,is__cell_barcode/\" &gt; CTRL_8_LA_singlecell.csv'</li><li>'cat D9_LA_per_barcode_metrics.csv | cut -d, -f1,4 | sed \"1s/.*/barcode,is__cell_barcode/\" &gt; D9_LA_singlecell.csv'</li><li>'cat D1_LA_per_barcode_metrics.csv | cut -d, -f1,4 | sed \"1s/.*/barcode,is__cell_barcode/\" &gt; D1_LA_singlecell.csv'</li><li>'cat HF6_LA_per_barcode_metrics.csv | cut -d, -f1,4 | sed \"1s/.*/barcode,is__cell_barcode/\" &gt; HF6_LA_singlecell.csv'</li><li>'cat HF9_LA_per_barcode_metrics.csv | cut -d, -f1,4 | sed \"1s/.*/barcode,is__cell_barcode/\" &gt; HF9_LA_singlecell.csv'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'cat AF\\_1\\_LA\\_per\\_barcode\\_metrics.csv \\textbar{} cut -d, -f1,4 \\textbar{} sed \"1s/.*/barcode,is\\_\\_cell\\_barcode/\" > AF\\_1\\_LA\\_singlecell.csv'\n",
       "\\item 'cat AF\\_2\\_LA\\_per\\_barcode\\_metrics.csv \\textbar{} cut -d, -f1,4 \\textbar{} sed \"1s/.*/barcode,is\\_\\_cell\\_barcode/\" > AF\\_2\\_LA\\_singlecell.csv'\n",
       "\\item 'cat AF\\_3\\_LA\\_per\\_barcode\\_metrics.csv \\textbar{} cut -d, -f1,4 \\textbar{} sed \"1s/.*/barcode,is\\_\\_cell\\_barcode/\" > AF\\_3\\_LA\\_singlecell.csv'\n",
       "\\item 'cat AF\\_4\\_LA\\_per\\_barcode\\_metrics.csv \\textbar{} cut -d, -f1,4 \\textbar{} sed \"1s/.*/barcode,is\\_\\_cell\\_barcode/\" > AF\\_4\\_LA\\_singlecell.csv'\n",
       "\\item 'cat AF\\_5\\_LA\\_per\\_barcode\\_metrics.csv \\textbar{} cut -d, -f1,4 \\textbar{} sed \"1s/.*/barcode,is\\_\\_cell\\_barcode/\" > AF\\_5\\_LA\\_singlecell.csv'\n",
       "\\item 'cat AF\\_6\\_LA\\_per\\_barcode\\_metrics.csv \\textbar{} cut -d, -f1,4 \\textbar{} sed \"1s/.*/barcode,is\\_\\_cell\\_barcode/\" > AF\\_6\\_LA\\_singlecell.csv'\n",
       "\\item 'cat AF\\_7\\_LA\\_per\\_barcode\\_metrics.csv \\textbar{} cut -d, -f1,4 \\textbar{} sed \"1s/.*/barcode,is\\_\\_cell\\_barcode/\" > AF\\_7\\_LA\\_singlecell.csv'\n",
       "\\item 'cat AF\\_8\\_LA\\_per\\_barcode\\_metrics.csv \\textbar{} cut -d, -f1,4 \\textbar{} sed \"1s/.*/barcode,is\\_\\_cell\\_barcode/\" > AF\\_8\\_LA\\_singlecell.csv'\n",
       "\\item 'cat CTRL\\_1\\_LA\\_per\\_barcode\\_metrics.csv \\textbar{} cut -d, -f1,4 \\textbar{} sed \"1s/.*/barcode,is\\_\\_cell\\_barcode/\" > CTRL\\_1\\_LA\\_singlecell.csv'\n",
       "\\item 'cat CTRL\\_2\\_LA\\_per\\_barcode\\_metrics.csv \\textbar{} cut -d, -f1,4 \\textbar{} sed \"1s/.*/barcode,is\\_\\_cell\\_barcode/\" > CTRL\\_2\\_LA\\_singlecell.csv'\n",
       "\\item 'cat CTRL\\_3\\_LA\\_per\\_barcode\\_metrics.csv \\textbar{} cut -d, -f1,4 \\textbar{} sed \"1s/.*/barcode,is\\_\\_cell\\_barcode/\" > CTRL\\_3\\_LA\\_singlecell.csv'\n",
       "\\item 'cat CTRL\\_4\\_LA\\_per\\_barcode\\_metrics.csv \\textbar{} cut -d, -f1,4 \\textbar{} sed \"1s/.*/barcode,is\\_\\_cell\\_barcode/\" > CTRL\\_4\\_LA\\_singlecell.csv'\n",
       "\\item 'cat CTRL\\_5\\_LA\\_per\\_barcode\\_metrics.csv \\textbar{} cut -d, -f1,4 \\textbar{} sed \"1s/.*/barcode,is\\_\\_cell\\_barcode/\" > CTRL\\_5\\_LA\\_singlecell.csv'\n",
       "\\item 'cat CTRL\\_6\\_LA\\_per\\_barcode\\_metrics.csv \\textbar{} cut -d, -f1,4 \\textbar{} sed \"1s/.*/barcode,is\\_\\_cell\\_barcode/\" > CTRL\\_6\\_LA\\_singlecell.csv'\n",
       "\\item 'cat CTRL\\_7\\_LA\\_per\\_barcode\\_metrics.csv \\textbar{} cut -d, -f1,4 \\textbar{} sed \"1s/.*/barcode,is\\_\\_cell\\_barcode/\" > CTRL\\_7\\_LA\\_singlecell.csv'\n",
       "\\item 'cat CTRL\\_8\\_LA\\_per\\_barcode\\_metrics.csv \\textbar{} cut -d, -f1,4 \\textbar{} sed \"1s/.*/barcode,is\\_\\_cell\\_barcode/\" > CTRL\\_8\\_LA\\_singlecell.csv'\n",
       "\\item 'cat D9\\_LA\\_per\\_barcode\\_metrics.csv \\textbar{} cut -d, -f1,4 \\textbar{} sed \"1s/.*/barcode,is\\_\\_cell\\_barcode/\" > D9\\_LA\\_singlecell.csv'\n",
       "\\item 'cat D1\\_LA\\_per\\_barcode\\_metrics.csv \\textbar{} cut -d, -f1,4 \\textbar{} sed \"1s/.*/barcode,is\\_\\_cell\\_barcode/\" > D1\\_LA\\_singlecell.csv'\n",
       "\\item 'cat HF6\\_LA\\_per\\_barcode\\_metrics.csv \\textbar{} cut -d, -f1,4 \\textbar{} sed \"1s/.*/barcode,is\\_\\_cell\\_barcode/\" > HF6\\_LA\\_singlecell.csv'\n",
       "\\item 'cat HF9\\_LA\\_per\\_barcode\\_metrics.csv \\textbar{} cut -d, -f1,4 \\textbar{} sed \"1s/.*/barcode,is\\_\\_cell\\_barcode/\" > HF9\\_LA\\_singlecell.csv'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'cat AF_1_LA_per_barcode_metrics.csv | cut -d, -f1,4 | sed \"1s/.*/barcode,is__cell_barcode/\" &gt; AF_1_LA_singlecell.csv'\n",
       "2. 'cat AF_2_LA_per_barcode_metrics.csv | cut -d, -f1,4 | sed \"1s/.*/barcode,is__cell_barcode/\" &gt; AF_2_LA_singlecell.csv'\n",
       "3. 'cat AF_3_LA_per_barcode_metrics.csv | cut -d, -f1,4 | sed \"1s/.*/barcode,is__cell_barcode/\" &gt; AF_3_LA_singlecell.csv'\n",
       "4. 'cat AF_4_LA_per_barcode_metrics.csv | cut -d, -f1,4 | sed \"1s/.*/barcode,is__cell_barcode/\" &gt; AF_4_LA_singlecell.csv'\n",
       "5. 'cat AF_5_LA_per_barcode_metrics.csv | cut -d, -f1,4 | sed \"1s/.*/barcode,is__cell_barcode/\" &gt; AF_5_LA_singlecell.csv'\n",
       "6. 'cat AF_6_LA_per_barcode_metrics.csv | cut -d, -f1,4 | sed \"1s/.*/barcode,is__cell_barcode/\" &gt; AF_6_LA_singlecell.csv'\n",
       "7. 'cat AF_7_LA_per_barcode_metrics.csv | cut -d, -f1,4 | sed \"1s/.*/barcode,is__cell_barcode/\" &gt; AF_7_LA_singlecell.csv'\n",
       "8. 'cat AF_8_LA_per_barcode_metrics.csv | cut -d, -f1,4 | sed \"1s/.*/barcode,is__cell_barcode/\" &gt; AF_8_LA_singlecell.csv'\n",
       "9. 'cat CTRL_1_LA_per_barcode_metrics.csv | cut -d, -f1,4 | sed \"1s/.*/barcode,is__cell_barcode/\" &gt; CTRL_1_LA_singlecell.csv'\n",
       "10. 'cat CTRL_2_LA_per_barcode_metrics.csv | cut -d, -f1,4 | sed \"1s/.*/barcode,is__cell_barcode/\" &gt; CTRL_2_LA_singlecell.csv'\n",
       "11. 'cat CTRL_3_LA_per_barcode_metrics.csv | cut -d, -f1,4 | sed \"1s/.*/barcode,is__cell_barcode/\" &gt; CTRL_3_LA_singlecell.csv'\n",
       "12. 'cat CTRL_4_LA_per_barcode_metrics.csv | cut -d, -f1,4 | sed \"1s/.*/barcode,is__cell_barcode/\" &gt; CTRL_4_LA_singlecell.csv'\n",
       "13. 'cat CTRL_5_LA_per_barcode_metrics.csv | cut -d, -f1,4 | sed \"1s/.*/barcode,is__cell_barcode/\" &gt; CTRL_5_LA_singlecell.csv'\n",
       "14. 'cat CTRL_6_LA_per_barcode_metrics.csv | cut -d, -f1,4 | sed \"1s/.*/barcode,is__cell_barcode/\" &gt; CTRL_6_LA_singlecell.csv'\n",
       "15. 'cat CTRL_7_LA_per_barcode_metrics.csv | cut -d, -f1,4 | sed \"1s/.*/barcode,is__cell_barcode/\" &gt; CTRL_7_LA_singlecell.csv'\n",
       "16. 'cat CTRL_8_LA_per_barcode_metrics.csv | cut -d, -f1,4 | sed \"1s/.*/barcode,is__cell_barcode/\" &gt; CTRL_8_LA_singlecell.csv'\n",
       "17. 'cat D9_LA_per_barcode_metrics.csv | cut -d, -f1,4 | sed \"1s/.*/barcode,is__cell_barcode/\" &gt; D9_LA_singlecell.csv'\n",
       "18. 'cat D1_LA_per_barcode_metrics.csv | cut -d, -f1,4 | sed \"1s/.*/barcode,is__cell_barcode/\" &gt; D1_LA_singlecell.csv'\n",
       "19. 'cat HF6_LA_per_barcode_metrics.csv | cut -d, -f1,4 | sed \"1s/.*/barcode,is__cell_barcode/\" &gt; HF6_LA_singlecell.csv'\n",
       "20. 'cat HF9_LA_per_barcode_metrics.csv | cut -d, -f1,4 | sed \"1s/.*/barcode,is__cell_barcode/\" &gt; HF9_LA_singlecell.csv'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"cat AF_1_LA_per_barcode_metrics.csv | cut -d, -f1,4 | sed \\\"1s/.*/barcode,is__cell_barcode/\\\" > AF_1_LA_singlecell.csv\"    \n",
       " [2] \"cat AF_2_LA_per_barcode_metrics.csv | cut -d, -f1,4 | sed \\\"1s/.*/barcode,is__cell_barcode/\\\" > AF_2_LA_singlecell.csv\"    \n",
       " [3] \"cat AF_3_LA_per_barcode_metrics.csv | cut -d, -f1,4 | sed \\\"1s/.*/barcode,is__cell_barcode/\\\" > AF_3_LA_singlecell.csv\"    \n",
       " [4] \"cat AF_4_LA_per_barcode_metrics.csv | cut -d, -f1,4 | sed \\\"1s/.*/barcode,is__cell_barcode/\\\" > AF_4_LA_singlecell.csv\"    \n",
       " [5] \"cat AF_5_LA_per_barcode_metrics.csv | cut -d, -f1,4 | sed \\\"1s/.*/barcode,is__cell_barcode/\\\" > AF_5_LA_singlecell.csv\"    \n",
       " [6] \"cat AF_6_LA_per_barcode_metrics.csv | cut -d, -f1,4 | sed \\\"1s/.*/barcode,is__cell_barcode/\\\" > AF_6_LA_singlecell.csv\"    \n",
       " [7] \"cat AF_7_LA_per_barcode_metrics.csv | cut -d, -f1,4 | sed \\\"1s/.*/barcode,is__cell_barcode/\\\" > AF_7_LA_singlecell.csv\"    \n",
       " [8] \"cat AF_8_LA_per_barcode_metrics.csv | cut -d, -f1,4 | sed \\\"1s/.*/barcode,is__cell_barcode/\\\" > AF_8_LA_singlecell.csv\"    \n",
       " [9] \"cat CTRL_1_LA_per_barcode_metrics.csv | cut -d, -f1,4 | sed \\\"1s/.*/barcode,is__cell_barcode/\\\" > CTRL_1_LA_singlecell.csv\"\n",
       "[10] \"cat CTRL_2_LA_per_barcode_metrics.csv | cut -d, -f1,4 | sed \\\"1s/.*/barcode,is__cell_barcode/\\\" > CTRL_2_LA_singlecell.csv\"\n",
       "[11] \"cat CTRL_3_LA_per_barcode_metrics.csv | cut -d, -f1,4 | sed \\\"1s/.*/barcode,is__cell_barcode/\\\" > CTRL_3_LA_singlecell.csv\"\n",
       "[12] \"cat CTRL_4_LA_per_barcode_metrics.csv | cut -d, -f1,4 | sed \\\"1s/.*/barcode,is__cell_barcode/\\\" > CTRL_4_LA_singlecell.csv\"\n",
       "[13] \"cat CTRL_5_LA_per_barcode_metrics.csv | cut -d, -f1,4 | sed \\\"1s/.*/barcode,is__cell_barcode/\\\" > CTRL_5_LA_singlecell.csv\"\n",
       "[14] \"cat CTRL_6_LA_per_barcode_metrics.csv | cut -d, -f1,4 | sed \\\"1s/.*/barcode,is__cell_barcode/\\\" > CTRL_6_LA_singlecell.csv\"\n",
       "[15] \"cat CTRL_7_LA_per_barcode_metrics.csv | cut -d, -f1,4 | sed \\\"1s/.*/barcode,is__cell_barcode/\\\" > CTRL_7_LA_singlecell.csv\"\n",
       "[16] \"cat CTRL_8_LA_per_barcode_metrics.csv | cut -d, -f1,4 | sed \\\"1s/.*/barcode,is__cell_barcode/\\\" > CTRL_8_LA_singlecell.csv\"\n",
       "[17] \"cat D9_LA_per_barcode_metrics.csv | cut -d, -f1,4 | sed \\\"1s/.*/barcode,is__cell_barcode/\\\" > D9_LA_singlecell.csv\"        \n",
       "[18] \"cat D1_LA_per_barcode_metrics.csv | cut -d, -f1,4 | sed \\\"1s/.*/barcode,is__cell_barcode/\\\" > D1_LA_singlecell.csv\"        \n",
       "[19] \"cat HF6_LA_per_barcode_metrics.csv | cut -d, -f1,4 | sed \\\"1s/.*/barcode,is__cell_barcode/\\\" > HF6_LA_singlecell.csv\"      \n",
       "[20] \"cat HF9_LA_per_barcode_metrics.csv | cut -d, -f1,4 | sed \\\"1s/.*/barcode,is__cell_barcode/\\\" > HF9_LA_singlecell.csv\"      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "paste0('cat ',sampleNames,'_per_barcode_metrics.csv | cut -d, -f1,4 | sed \"1s/.*/barcode,is__cell_barcode/\" > ',sampleNames,'_singlecell.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "4354fe0e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'/volume/afib/data/'"
      ],
      "text/latex": [
       "'/volume/afib/data/'"
      ],
      "text/markdown": [
       "'/volume/afib/data/'"
      ],
      "text/plain": [
       "[1] \"/volume/afib/data/\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "inputFiles_dir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "cf647d8c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"bash AMULET.sh /volume/afib/data/AF_1_LA_atac_fragments.tsv.gz /volume/afib/data/AF_1_LA_singlecell.csv /home/ubuntu/scripts/AMULET/human_autosomes.txt /home/ubuntu/scripts/AMULET/blacklist_repeats_segdups_rmsk_hg38.bed /volume/afib/out/QC_preprocessing/AF_1_LA_amulet/ /home/ubuntu/scripts/AMULET/\"\n",
      "[1] \"bash AMULET.sh /volume/afib/data/AF_2_LA_atac_fragments.tsv.gz /volume/afib/data/AF_2_LA_singlecell.csv /home/ubuntu/scripts/AMULET/human_autosomes.txt /home/ubuntu/scripts/AMULET/blacklist_repeats_segdups_rmsk_hg38.bed /volume/afib/out/QC_preprocessing/AF_2_LA_amulet/ /home/ubuntu/scripts/AMULET/\"\n",
      "[1] \"bash AMULET.sh /volume/afib/data/AF_3_LA_atac_fragments.tsv.gz /volume/afib/data/AF_3_LA_singlecell.csv /home/ubuntu/scripts/AMULET/human_autosomes.txt /home/ubuntu/scripts/AMULET/blacklist_repeats_segdups_rmsk_hg38.bed /volume/afib/out/QC_preprocessing/AF_3_LA_amulet/ /home/ubuntu/scripts/AMULET/\"\n",
      "[1] \"bash AMULET.sh /volume/afib/data/AF_4_LA_atac_fragments.tsv.gz /volume/afib/data/AF_4_LA_singlecell.csv /home/ubuntu/scripts/AMULET/human_autosomes.txt /home/ubuntu/scripts/AMULET/blacklist_repeats_segdups_rmsk_hg38.bed /volume/afib/out/QC_preprocessing/AF_4_LA_amulet/ /home/ubuntu/scripts/AMULET/\"\n",
      "[1] \"bash AMULET.sh /volume/afib/data/AF_5_LA_atac_fragments.tsv.gz /volume/afib/data/AF_5_LA_singlecell.csv /home/ubuntu/scripts/AMULET/human_autosomes.txt /home/ubuntu/scripts/AMULET/blacklist_repeats_segdups_rmsk_hg38.bed /volume/afib/out/QC_preprocessing/AF_5_LA_amulet/ /home/ubuntu/scripts/AMULET/\"\n",
      "[1] \"bash AMULET.sh /volume/afib/data/AF_6_LA_atac_fragments.tsv.gz /volume/afib/data/AF_6_LA_singlecell.csv /home/ubuntu/scripts/AMULET/human_autosomes.txt /home/ubuntu/scripts/AMULET/blacklist_repeats_segdups_rmsk_hg38.bed /volume/afib/out/QC_preprocessing/AF_6_LA_amulet/ /home/ubuntu/scripts/AMULET/\"\n",
      "[1] \"bash AMULET.sh /volume/afib/data/AF_7_LA_atac_fragments.tsv.gz /volume/afib/data/AF_7_LA_singlecell.csv /home/ubuntu/scripts/AMULET/human_autosomes.txt /home/ubuntu/scripts/AMULET/blacklist_repeats_segdups_rmsk_hg38.bed /volume/afib/out/QC_preprocessing/AF_7_LA_amulet/ /home/ubuntu/scripts/AMULET/\"\n",
      "[1] \"bash AMULET.sh /volume/afib/data/AF_8_LA_atac_fragments.tsv.gz /volume/afib/data/AF_8_LA_singlecell.csv /home/ubuntu/scripts/AMULET/human_autosomes.txt /home/ubuntu/scripts/AMULET/blacklist_repeats_segdups_rmsk_hg38.bed /volume/afib/out/QC_preprocessing/AF_8_LA_amulet/ /home/ubuntu/scripts/AMULET/\"\n",
      "[1] \"bash AMULET.sh /volume/afib/data/CTRL_1_LA_atac_fragments.tsv.gz /volume/afib/data/CTRL_1_LA_singlecell.csv /home/ubuntu/scripts/AMULET/human_autosomes.txt /home/ubuntu/scripts/AMULET/blacklist_repeats_segdups_rmsk_hg38.bed /volume/afib/out/QC_preprocessing/CTRL_1_LA_amulet/ /home/ubuntu/scripts/AMULET/\"\n",
      "[1] \"bash AMULET.sh /volume/afib/data/CTRL_2_LA_atac_fragments.tsv.gz /volume/afib/data/CTRL_2_LA_singlecell.csv /home/ubuntu/scripts/AMULET/human_autosomes.txt /home/ubuntu/scripts/AMULET/blacklist_repeats_segdups_rmsk_hg38.bed /volume/afib/out/QC_preprocessing/CTRL_2_LA_amulet/ /home/ubuntu/scripts/AMULET/\"\n",
      "[1] \"bash AMULET.sh /volume/afib/data/CTRL_3_LA_atac_fragments.tsv.gz /volume/afib/data/CTRL_3_LA_singlecell.csv /home/ubuntu/scripts/AMULET/human_autosomes.txt /home/ubuntu/scripts/AMULET/blacklist_repeats_segdups_rmsk_hg38.bed /volume/afib/out/QC_preprocessing/CTRL_3_LA_amulet/ /home/ubuntu/scripts/AMULET/\"\n",
      "[1] \"bash AMULET.sh /volume/afib/data/CTRL_4_LA_atac_fragments.tsv.gz /volume/afib/data/CTRL_4_LA_singlecell.csv /home/ubuntu/scripts/AMULET/human_autosomes.txt /home/ubuntu/scripts/AMULET/blacklist_repeats_segdups_rmsk_hg38.bed /volume/afib/out/QC_preprocessing/CTRL_4_LA_amulet/ /home/ubuntu/scripts/AMULET/\"\n",
      "[1] \"bash AMULET.sh /volume/afib/data/CTRL_5_LA_atac_fragments.tsv.gz /volume/afib/data/CTRL_5_LA_singlecell.csv /home/ubuntu/scripts/AMULET/human_autosomes.txt /home/ubuntu/scripts/AMULET/blacklist_repeats_segdups_rmsk_hg38.bed /volume/afib/out/QC_preprocessing/CTRL_5_LA_amulet/ /home/ubuntu/scripts/AMULET/\"\n",
      "[1] \"bash AMULET.sh /volume/afib/data/CTRL_6_LA_atac_fragments.tsv.gz /volume/afib/data/CTRL_6_LA_singlecell.csv /home/ubuntu/scripts/AMULET/human_autosomes.txt /home/ubuntu/scripts/AMULET/blacklist_repeats_segdups_rmsk_hg38.bed /volume/afib/out/QC_preprocessing/CTRL_6_LA_amulet/ /home/ubuntu/scripts/AMULET/\"\n",
      "[1] \"bash AMULET.sh /volume/afib/data/CTRL_7_LA_atac_fragments.tsv.gz /volume/afib/data/CTRL_7_LA_singlecell.csv /home/ubuntu/scripts/AMULET/human_autosomes.txt /home/ubuntu/scripts/AMULET/blacklist_repeats_segdups_rmsk_hg38.bed /volume/afib/out/QC_preprocessing/CTRL_7_LA_amulet/ /home/ubuntu/scripts/AMULET/\"\n",
      "[1] \"bash AMULET.sh /volume/afib/data/CTRL_8_LA_atac_fragments.tsv.gz /volume/afib/data/CTRL_8_LA_singlecell.csv /home/ubuntu/scripts/AMULET/human_autosomes.txt /home/ubuntu/scripts/AMULET/blacklist_repeats_segdups_rmsk_hg38.bed /volume/afib/out/QC_preprocessing/CTRL_8_LA_amulet/ /home/ubuntu/scripts/AMULET/\"\n",
      "[1] \"bash AMULET.sh /volume/afib/data/D9_LA_atac_fragments.tsv.gz /volume/afib/data/D9_LA_singlecell.csv /home/ubuntu/scripts/AMULET/human_autosomes.txt /home/ubuntu/scripts/AMULET/blacklist_repeats_segdups_rmsk_hg38.bed /volume/afib/out/QC_preprocessing/D9_LA_amulet/ /home/ubuntu/scripts/AMULET/\"\n",
      "[1] \"bash AMULET.sh /volume/afib/data/D1_LA_atac_fragments.tsv.gz /volume/afib/data/D1_LA_singlecell.csv /home/ubuntu/scripts/AMULET/human_autosomes.txt /home/ubuntu/scripts/AMULET/blacklist_repeats_segdups_rmsk_hg38.bed /volume/afib/out/QC_preprocessing/D1_LA_amulet/ /home/ubuntu/scripts/AMULET/\"\n",
      "[1] \"bash AMULET.sh /volume/afib/data/HF6_LA_atac_fragments.tsv.gz /volume/afib/data/HF6_LA_singlecell.csv /home/ubuntu/scripts/AMULET/human_autosomes.txt /home/ubuntu/scripts/AMULET/blacklist_repeats_segdups_rmsk_hg38.bed /volume/afib/out/QC_preprocessing/HF6_LA_amulet/ /home/ubuntu/scripts/AMULET/\"\n",
      "[1] \"bash AMULET.sh /volume/afib/data/HF9_LA_atac_fragments.tsv.gz /volume/afib/data/HF9_LA_singlecell.csv /home/ubuntu/scripts/AMULET/human_autosomes.txt /home/ubuntu/scripts/AMULET/blacklist_repeats_segdups_rmsk_hg38.bed /volume/afib/out/QC_preprocessing/HF9_LA_amulet/ /home/ubuntu/scripts/AMULET/\"\n"
     ]
    }
   ],
   "source": [
    "# create linux folders for the AMULET outputs if the folder doesn't already exist and write out the AMULET command line\n",
    "for (i in seq_along(filtered_list)){\n",
    "    sampleName <- names(filtered_list[i])\n",
    "    #data_folder <- inputFiles_dir[[sampleName]]\n",
    "    folder <- paste0(out_dir,sampleName,\"_amulet\")\n",
    "    \n",
    "    if (!(dir.exists(folder))) {\n",
    "        dir.create(folder)\n",
    "    }\n",
    "    \n",
    "    print(paste0('bash AMULET.sh ',inputFiles_dir,sampleName,'_atac_fragments.tsv.gz ',\n",
    "                 inputFiles_dir,sampleName,'_singlecell.csv ',\n",
    "                 '/home/ubuntu/scripts/AMULET/human_autosomes.txt ',\n",
    "                 '/home/ubuntu/scripts/AMULET/blacklist_repeats_segdups_rmsk_hg38.bed ',\n",
    "                 folder,'/ ',\n",
    "                 '/home/ubuntu/scripts/AMULET/'\n",
    "                 )\n",
    "         )\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "721deb67",
   "metadata": {},
   "source": [
    "# Remake Seurat objects on scrublet + amulet corrected data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a724bc2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load in the postQCfilt.rds objects and remove the doublets from scrublet and amulet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "797da1b3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'/volume/afib/out/QC_preprocessing/'"
      ],
      "text/latex": [
       "'/volume/afib/out/QC\\_preprocessing/'"
      ],
      "text/markdown": [
       "'/volume/afib/out/QC_preprocessing/'"
      ],
      "text/plain": [
       "[1] \"/volume/afib/out/QC_preprocessing/\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "out_dir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b0811d49",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'AF_1_LA'</li><li>'AF_2_LA'</li><li>'AF_3_LA'</li><li>'AF_4_LA'</li><li>'AF_5_LA'</li><li>'AF_6_LA'</li><li>'AF_7_LA'</li><li>'AF_8_LA'</li><li>'CTRL_1_LA'</li><li>'CTRL_2_LA'</li><li>'CTRL_3_LA'</li><li>'CTRL_4_LA'</li><li>'CTRL_5_LA'</li><li>'CTRL_6_LA'</li><li>'CTRL_7_LA'</li><li>'CTRL_8_LA'</li><li>'D9_LA'</li><li>'D1_LA'</li><li>'HF6_LA'</li><li>'HF9_LA'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'AF\\_1\\_LA'\n",
       "\\item 'AF\\_2\\_LA'\n",
       "\\item 'AF\\_3\\_LA'\n",
       "\\item 'AF\\_4\\_LA'\n",
       "\\item 'AF\\_5\\_LA'\n",
       "\\item 'AF\\_6\\_LA'\n",
       "\\item 'AF\\_7\\_LA'\n",
       "\\item 'AF\\_8\\_LA'\n",
       "\\item 'CTRL\\_1\\_LA'\n",
       "\\item 'CTRL\\_2\\_LA'\n",
       "\\item 'CTRL\\_3\\_LA'\n",
       "\\item 'CTRL\\_4\\_LA'\n",
       "\\item 'CTRL\\_5\\_LA'\n",
       "\\item 'CTRL\\_6\\_LA'\n",
       "\\item 'CTRL\\_7\\_LA'\n",
       "\\item 'CTRL\\_8\\_LA'\n",
       "\\item 'D9\\_LA'\n",
       "\\item 'D1\\_LA'\n",
       "\\item 'HF6\\_LA'\n",
       "\\item 'HF9\\_LA'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'AF_1_LA'\n",
       "2. 'AF_2_LA'\n",
       "3. 'AF_3_LA'\n",
       "4. 'AF_4_LA'\n",
       "5. 'AF_5_LA'\n",
       "6. 'AF_6_LA'\n",
       "7. 'AF_7_LA'\n",
       "8. 'AF_8_LA'\n",
       "9. 'CTRL_1_LA'\n",
       "10. 'CTRL_2_LA'\n",
       "11. 'CTRL_3_LA'\n",
       "12. 'CTRL_4_LA'\n",
       "13. 'CTRL_5_LA'\n",
       "14. 'CTRL_6_LA'\n",
       "15. 'CTRL_7_LA'\n",
       "16. 'CTRL_8_LA'\n",
       "17. 'D9_LA'\n",
       "18. 'D1_LA'\n",
       "19. 'HF6_LA'\n",
       "20. 'HF9_LA'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"AF_1_LA\"   \"AF_2_LA\"   \"AF_3_LA\"   \"AF_4_LA\"   \"AF_5_LA\"   \"AF_6_LA\"  \n",
       " [7] \"AF_7_LA\"   \"AF_8_LA\"   \"CTRL_1_LA\" \"CTRL_2_LA\" \"CTRL_3_LA\" \"CTRL_4_LA\"\n",
       "[13] \"CTRL_5_LA\" \"CTRL_6_LA\" \"CTRL_7_LA\" \"CTRL_8_LA\" \"D9_LA\"     \"D1_LA\"    \n",
       "[19] \"HF6_LA\"    \"HF9_LA\"   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sampleNames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "04ecf3ac",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'/volume/afib/data/'"
      ],
      "text/latex": [
       "'/volume/afib/data/'"
      ],
      "text/markdown": [
       "'/volume/afib/data/'"
      ],
      "text/plain": [
       "[1] \"/volume/afib/data/\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "inputFiles_dir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "33cd6755",
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered_list <- list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "f07b1340",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Removing doublets: AF_1_LA\"\n",
      "[1] \"scrublet results:\"\n",
      "\n",
      "doublet singlet \n",
      "    390    4672 \n",
      "[1] \"amulet results:\"\n",
      "\n",
      "doublet singlet \n",
      "    155    4907 \n",
      "[1] \"total doublets to remove (scrublet + amulet):\"\n",
      "\n",
      "doublet singlet \n",
      "    486    4576 \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Computing hash\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"After doublet filtering: AF_1_LA\"\n",
      "An object of class Seurat \n",
      "110887 features across 4576 samples within 2 assays \n",
      "Active assay: ATAC (74286 features, 0 variable features)\n",
      " 1 other assay present: RNA\n",
      " 5 dimensional reductions calculated: pca, umap.rna, lsi, umap.atac, umap.wnn\n",
      "[1] \"Removing doublets: AF_2_LA\"\n",
      "[1] \"scrublet results:\"\n",
      "\n",
      "doublet singlet \n",
      "    111    2526 \n",
      "[1] \"amulet results:\"\n",
      "\n",
      "doublet singlet \n",
      "     66    2571 \n",
      "[1] \"total doublets to remove (scrublet + amulet):\"\n",
      "\n",
      "doublet singlet \n",
      "    166    2471 \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Computing hash\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"After doublet filtering: AF_2_LA\"\n",
      "An object of class Seurat \n",
      "99767 features across 2471 samples within 2 assays \n",
      "Active assay: ATAC (63166 features, 0 variable features)\n",
      " 1 other assay present: RNA\n",
      " 5 dimensional reductions calculated: pca, umap.rna, lsi, umap.atac, umap.wnn\n",
      "[1] \"Removing doublets: AF_3_LA\"\n",
      "[1] \"scrublet results:\"\n",
      "\n",
      "doublet singlet \n",
      "    430    7743 \n",
      "[1] \"amulet results:\"\n",
      "\n",
      "doublet singlet \n",
      "    138    8035 \n",
      "[1] \"total doublets to remove (scrublet + amulet):\"\n",
      "\n",
      "doublet singlet \n",
      "    533    7640 \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Computing hash\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"After doublet filtering: AF_3_LA\"\n",
      "An object of class Seurat \n",
      "115532 features across 7640 samples within 2 assays \n",
      "Active assay: ATAC (78931 features, 0 variable features)\n",
      " 1 other assay present: RNA\n",
      " 5 dimensional reductions calculated: pca, umap.rna, lsi, umap.atac, umap.wnn\n",
      "[1] \"Removing doublets: AF_4_LA\"\n",
      "[1] \"scrublet results:\"\n",
      "\n",
      "doublet singlet \n",
      "    493    6339 \n",
      "[1] \"amulet results:\"\n",
      "\n",
      "doublet singlet \n",
      "    310    6522 \n",
      "[1] \"total doublets to remove (scrublet + amulet):\"\n",
      "\n",
      "doublet singlet \n",
      "    708    6124 \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Computing hash\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"After doublet filtering: AF_4_LA\"\n",
      "An object of class Seurat \n",
      "117744 features across 6124 samples within 2 assays \n",
      "Active assay: ATAC (81143 features, 0 variable features)\n",
      " 1 other assay present: RNA\n",
      " 5 dimensional reductions calculated: pca, umap.rna, lsi, umap.atac, umap.wnn\n",
      "[1] \"Removing doublets: AF_5_LA\"\n",
      "[1] \"scrublet results:\"\n",
      "\n",
      "doublet singlet \n",
      "   1141   12659 \n",
      "[1] \"amulet results:\"\n",
      "\n",
      "doublet singlet \n",
      "    127   13673 \n",
      "[1] \"total doublets to remove (scrublet + amulet):\"\n",
      "\n",
      "doublet singlet \n",
      "   1220   12580 \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Computing hash\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"After doublet filtering: AF_5_LA\"\n",
      "An object of class Seurat \n",
      "140002 features across 12580 samples within 2 assays \n",
      "Active assay: ATAC (103401 features, 0 variable features)\n",
      " 1 other assay present: RNA\n",
      " 5 dimensional reductions calculated: pca, umap.rna, lsi, umap.atac, umap.wnn\n",
      "[1] \"Removing doublets: AF_6_LA\"\n",
      "[1] \"scrublet results:\"\n",
      "\n",
      "doublet singlet \n",
      "    521    9846 \n",
      "[1] \"amulet results:\"\n",
      "\n",
      "doublet singlet \n",
      "    173   10194 \n",
      "[1] \"total doublets to remove (scrublet + amulet):\"\n",
      "\n",
      "doublet singlet \n",
      "    639    9728 \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Computing hash\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"After doublet filtering: AF_6_LA\"\n",
      "An object of class Seurat \n",
      "144509 features across 9728 samples within 2 assays \n",
      "Active assay: ATAC (107908 features, 0 variable features)\n",
      " 1 other assay present: RNA\n",
      " 5 dimensional reductions calculated: pca, umap.rna, lsi, umap.atac, umap.wnn\n",
      "[1] \"Removing doublets: AF_7_LA\"\n",
      "[1] \"scrublet results:\"\n",
      "\n",
      "doublet singlet \n",
      "    687    8764 \n",
      "[1] \"amulet results:\"\n",
      "\n",
      "doublet singlet \n",
      "    251    9200 \n",
      "[1] \"total doublets to remove (scrublet + amulet):\"\n",
      "\n",
      "doublet singlet \n",
      "    869    8582 \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Computing hash\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"After doublet filtering: AF_7_LA\"\n",
      "An object of class Seurat \n",
      "111618 features across 8582 samples within 2 assays \n",
      "Active assay: ATAC (75017 features, 0 variable features)\n",
      " 1 other assay present: RNA\n",
      " 5 dimensional reductions calculated: pca, umap.rna, lsi, umap.atac, umap.wnn\n",
      "[1] \"Removing doublets: AF_8_LA\"\n",
      "[1] \"scrublet results:\"\n",
      "\n",
      "doublet singlet \n",
      "    468    7302 \n",
      "[1] \"amulet results:\"\n",
      "\n",
      "doublet singlet \n",
      "     31    7739 \n",
      "[1] \"total doublets to remove (scrublet + amulet):\"\n",
      "\n",
      "doublet singlet \n",
      "    489    7281 \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Computing hash\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"After doublet filtering: AF_8_LA\"\n",
      "An object of class Seurat \n",
      "148797 features across 7281 samples within 2 assays \n",
      "Active assay: ATAC (112196 features, 0 variable features)\n",
      " 1 other assay present: RNA\n",
      " 5 dimensional reductions calculated: pca, umap.rna, lsi, umap.atac, umap.wnn\n",
      "[1] \"Removing doublets: CTRL_1_LA\"\n",
      "[1] \"scrublet results:\"\n",
      "\n",
      "doublet singlet \n",
      "    431    5930 \n",
      "[1] \"amulet results:\"\n",
      "\n",
      "doublet singlet \n",
      "     86    6275 \n",
      "[1] \"total doublets to remove (scrublet + amulet):\"\n",
      "\n",
      "doublet singlet \n",
      "    480    5881 \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Computing hash\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"After doublet filtering: CTRL_1_LA\"\n",
      "An object of class Seurat \n",
      "131905 features across 5881 samples within 2 assays \n",
      "Active assay: ATAC (95304 features, 0 variable features)\n",
      " 1 other assay present: RNA\n",
      " 5 dimensional reductions calculated: pca, umap.rna, lsi, umap.atac, umap.wnn\n",
      "[1] \"Removing doublets: CTRL_2_LA\"\n",
      "[1] \"scrublet results:\"\n",
      "\n",
      "doublet singlet \n",
      "    549    7532 \n",
      "[1] \"amulet results:\"\n",
      "\n",
      "doublet singlet \n",
      "    150    7931 \n",
      "[1] \"total doublets to remove (scrublet + amulet):\"\n",
      "\n",
      "doublet singlet \n",
      "    648    7433 \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Computing hash\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"After doublet filtering: CTRL_2_LA\"\n",
      "An object of class Seurat \n",
      "161621 features across 7433 samples within 2 assays \n",
      "Active assay: ATAC (125020 features, 0 variable features)\n",
      " 1 other assay present: RNA\n",
      " 5 dimensional reductions calculated: pca, umap.rna, lsi, umap.atac, umap.wnn\n",
      "[1] \"Removing doublets: CTRL_3_LA\"\n",
      "[1] \"scrublet results:\"\n",
      "\n",
      "doublet singlet \n",
      "    586    8964 \n",
      "[1] \"amulet results:\"\n",
      "\n",
      "doublet singlet \n",
      "     43    9507 \n",
      "[1] \"total doublets to remove (scrublet + amulet):\"\n",
      "\n",
      "doublet singlet \n",
      "    612    8938 \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Computing hash\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"After doublet filtering: CTRL_3_LA\"\n",
      "An object of class Seurat \n",
      "148085 features across 8938 samples within 2 assays \n",
      "Active assay: ATAC (111484 features, 0 variable features)\n",
      " 1 other assay present: RNA\n",
      " 5 dimensional reductions calculated: pca, umap.rna, lsi, umap.atac, umap.wnn\n",
      "[1] \"Removing doublets: CTRL_4_LA\"\n",
      "[1] \"scrublet results:\"\n",
      "\n",
      "doublet singlet \n",
      "    395    8082 \n",
      "[1] \"amulet results:\"\n",
      "\n",
      "doublet singlet \n",
      "    342    8135 \n",
      "[1] \"total doublets to remove (scrublet + amulet):\"\n",
      "\n",
      "doublet singlet \n",
      "    683    7794 \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Computing hash\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"After doublet filtering: CTRL_4_LA\"\n",
      "An object of class Seurat \n",
      "148274 features across 7794 samples within 2 assays \n",
      "Active assay: ATAC (111673 features, 0 variable features)\n",
      " 1 other assay present: RNA\n",
      " 5 dimensional reductions calculated: pca, umap.rna, lsi, umap.atac, umap.wnn\n",
      "[1] \"Removing doublets: CTRL_5_LA\"\n",
      "[1] \"scrublet results:\"\n",
      "\n",
      "doublet singlet \n",
      "    342    7219 \n",
      "[1] \"amulet results:\"\n",
      "\n",
      "doublet singlet \n",
      "    105    7456 \n",
      "[1] \"total doublets to remove (scrublet + amulet):\"\n",
      "\n",
      "doublet singlet \n",
      "    409    7152 \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Computing hash\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"After doublet filtering: CTRL_5_LA\"\n",
      "An object of class Seurat \n",
      "121935 features across 7152 samples within 2 assays \n",
      "Active assay: ATAC (85334 features, 0 variable features)\n",
      " 1 other assay present: RNA\n",
      " 5 dimensional reductions calculated: pca, umap.rna, lsi, umap.atac, umap.wnn\n",
      "[1] \"Removing doublets: CTRL_6_LA\"\n",
      "[1] \"scrublet results:\"\n",
      "\n",
      "doublet singlet \n",
      "    759    9809 \n",
      "[1] \"amulet results:\"\n",
      "\n",
      "doublet singlet \n",
      "     73   10495 \n",
      "[1] \"total doublets to remove (scrublet + amulet):\"\n",
      "\n",
      "doublet singlet \n",
      "    805    9763 \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Computing hash\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"After doublet filtering: CTRL_6_LA\"\n",
      "An object of class Seurat \n",
      "140232 features across 9763 samples within 2 assays \n",
      "Active assay: ATAC (103631 features, 0 variable features)\n",
      " 1 other assay present: RNA\n",
      " 5 dimensional reductions calculated: pca, umap.rna, lsi, umap.atac, umap.wnn\n",
      "[1] \"Removing doublets: CTRL_7_LA\"\n",
      "[1] \"scrublet results:\"\n",
      "\n",
      "doublet singlet \n",
      "    488    7193 \n",
      "[1] \"amulet results:\"\n",
      "\n",
      "doublet singlet \n",
      "     67    7614 \n",
      "[1] \"total doublets to remove (scrublet + amulet):\"\n",
      "\n",
      "doublet singlet \n",
      "    529    7152 \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Computing hash\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"After doublet filtering: CTRL_7_LA\"\n",
      "An object of class Seurat \n",
      "173298 features across 7152 samples within 2 assays \n",
      "Active assay: ATAC (136697 features, 0 variable features)\n",
      " 1 other assay present: RNA\n",
      " 5 dimensional reductions calculated: pca, umap.rna, lsi, umap.atac, umap.wnn\n",
      "[1] \"Removing doublets: CTRL_8_LA\"\n",
      "[1] \"scrublet results:\"\n",
      "\n",
      "doublet singlet \n",
      "    468    6095 \n",
      "[1] \"amulet results:\"\n",
      "\n",
      "doublet singlet \n",
      "     77    6486 \n",
      "[1] \"total doublets to remove (scrublet + amulet):\"\n",
      "\n",
      "doublet singlet \n",
      "    516    6047 \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Computing hash\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"After doublet filtering: CTRL_8_LA\"\n",
      "An object of class Seurat \n",
      "95270 features across 6047 samples within 2 assays \n",
      "Active assay: ATAC (58669 features, 0 variable features)\n",
      " 1 other assay present: RNA\n",
      " 5 dimensional reductions calculated: pca, umap.rna, lsi, umap.atac, umap.wnn\n",
      "[1] \"Removing doublets: D9_LA\"\n",
      "[1] \"scrublet results:\"\n",
      "\n",
      "doublet singlet \n",
      "    458    8125 \n",
      "[1] \"amulet results:\"\n",
      "\n",
      "doublet singlet \n",
      "    434    8149 \n",
      "[1] \"total doublets to remove (scrublet + amulet):\"\n",
      "\n",
      "doublet singlet \n",
      "    827    7756 \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Computing hash\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"After doublet filtering: D9_LA\"\n",
      "An object of class Seurat \n",
      "145798 features across 7756 samples within 2 assays \n",
      "Active assay: ATAC (109197 features, 0 variable features)\n",
      " 1 other assay present: RNA\n",
      " 5 dimensional reductions calculated: pca, umap.rna, lsi, umap.atac, umap.wnn\n",
      "[1] \"Removing doublets: D1_LA\"\n",
      "[1] \"scrublet results:\"\n",
      "\n",
      "doublet singlet \n",
      "    241    4063 \n",
      "[1] \"amulet results:\"\n",
      "\n",
      "doublet singlet \n",
      "    238    4066 \n",
      "[1] \"total doublets to remove (scrublet + amulet):\"\n",
      "\n",
      "doublet singlet \n",
      "    442    3862 \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Computing hash\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"After doublet filtering: D1_LA\"\n",
      "An object of class Seurat \n",
      "131885 features across 3862 samples within 2 assays \n",
      "Active assay: ATAC (95284 features, 0 variable features)\n",
      " 1 other assay present: RNA\n",
      " 5 dimensional reductions calculated: pca, umap.rna, lsi, umap.atac, umap.wnn\n",
      "[1] \"Removing doublets: HF6_LA\"\n",
      "[1] \"scrublet results:\"\n",
      "\n",
      "doublet singlet \n",
      "    492    8576 \n",
      "[1] \"amulet results:\"\n",
      "\n",
      "doublet singlet \n",
      "    476    8592 \n",
      "[1] \"total doublets to remove (scrublet + amulet):\"\n",
      "\n",
      "doublet singlet \n",
      "    868    8200 \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Computing hash\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"After doublet filtering: HF6_LA\"\n",
      "An object of class Seurat \n",
      "125162 features across 8200 samples within 2 assays \n",
      "Active assay: ATAC (88561 features, 0 variable features)\n",
      " 1 other assay present: RNA\n",
      " 5 dimensional reductions calculated: pca, umap.rna, lsi, umap.atac, umap.wnn\n",
      "[1] \"Removing doublets: HF9_LA\"\n",
      "[1] \"scrublet results:\"\n",
      "\n",
      "doublet singlet \n",
      "    350    7071 \n",
      "[1] \"amulet results:\"\n",
      "\n",
      "doublet singlet \n",
      "    237    7184 \n",
      "[1] \"total doublets to remove (scrublet + amulet):\"\n",
      "\n",
      "doublet singlet \n",
      "    521    6900 \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Computing hash\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"After doublet filtering: HF9_LA\"\n",
      "An object of class Seurat \n",
      "137040 features across 6900 samples within 2 assays \n",
      "Active assay: ATAC (100439 features, 0 variable features)\n",
      " 1 other assay present: RNA\n",
      " 5 dimensional reductions calculated: pca, umap.rna, lsi, umap.atac, umap.wnn\n"
     ]
    }
   ],
   "source": [
    "for (i in seq_along(sampleNames)){\n",
    "    sampleName <- sampleNames[i]\n",
    "    seurat <- readRDS(paste0(out_dir,sampleName,'_postQCfilt.rds'))\n",
    "    #sampleFile_rna <- paste0(inputFiles_dir[[i]],sampleName,'_soupX/')\n",
    "    sampleFile_atac <- paste0(inputFiles_dir,sampleName,'_atac_fragments.tsv.gz')\n",
    "    print(paste0('Removing doublets: ', sampleName))\n",
    "    \n",
    "    # Scrublet\n",
    "    # Load in scrublet results after running the python script on soupx counts and add to metadata\n",
    "    scrublet <- read.table(paste0(out_dir,sampleName,'_scrublet/scrublet_results.tsv'), sep='\\t', header=TRUE)\n",
    "    print('scrublet results:')\n",
    "    print(table(scrublet$scrublet_DropletType))\n",
    "    # seurat metadata cellnames already in the same order as scrublet result cellnames so we can just add the metadata in\n",
    "    seurat@meta.data$scrublet_DropletType <- scrublet$scrublet_DropletType\n",
    "    seurat@meta.data$scrublet_Scores <- scrublet$scrublet_Scores\n",
    "    \n",
    "    # AMULET\n",
    "    # Load in Amulet results after running the bash script on celranger fragments and add to metadata\n",
    "    amulet <- read.table(paste0(out_dir,sampleName,'_amulet/MultipletBarcodes_01.txt'), sep='\\t', header=FALSE)\n",
    "    seurat@meta.data$amulet_doubletType <- 'singlet'\n",
    "    for (cell in amulet$V1) {\n",
    "        if (cell %in% rownames(seurat@meta.data)) { # need to check if cell exists in our seurat object because amulet was run on all fragments from cellranger output\n",
    "            seurat@meta.data[cell,'amulet_doubletType'] <- 'doublet'\n",
    "        }\n",
    "    }\n",
    "    print('amulet results:')\n",
    "    print(table(seurat@meta.data$amulet_doubletType))\n",
    "    \n",
    "    # create all_doublets metadata for both scrublet+amulet\n",
    "    # get the cells that were deemed a doublet by scrublet or amulet\n",
    "    cells <- rownames(seurat@meta.data %>% dplyr::filter(scrublet_DropletType == 'doublet' | amulet_doubletType == 'doublet'))\n",
    "    seurat@meta.data$all_doublets <- 'singlet'\n",
    "    for (cell in cells) {\n",
    "        if (cell %in% rownames(seurat@meta.data)) { # need to check if cell exists in our seurat object (even though it is, good practice to check)\n",
    "            seurat@meta.data[cell,'all_doublets'] <- 'doublet'\n",
    "        }\n",
    "    }\n",
    "    print('total doublets to remove (scrublet + amulet):')\n",
    "    print(table(seurat@meta.data$all_doublets))\n",
    "    \n",
    "    pdf(paste0(out_dir,'UMAPs_',sampleName,'_doublets.pdf'), width=7, height=7)\n",
    "    print(DimPlot(seurat, group.by='scrublet_DropletType', reduction = \"umap.rna\"))\n",
    "    print(DimPlot(seurat, group.by='amulet_doubletType', reduction = \"umap.atac\"))\n",
    "    print(DimPlot(seurat, group.by='all_doublets', reduction = \"umap.wnn\"))\n",
    "    dev.off()\n",
    "       \n",
    "    # Filter out the doublets\n",
    "    Idents(seurat) <- 'all_doublets'\n",
    "    seurat_filtered <- subset(seurat, idents='singlet')\n",
    "        \n",
    "    # Filter atac counts matrix to the filtered cells and remake the chromatin assay (possibly unnecessary, not sure)\n",
    "    atac_counts <- seurat_filtered@assays$ATAC@counts\n",
    "    #atac_counts <- atac_counts[,colnames(seurat_filtered)]\n",
    "    \n",
    "    grange.counts <- StringToGRanges(rownames(atac_counts), sep = c(\":\", \"-\"))\n",
    "    grange.use <- seqnames(grange.counts) %in% standardChromosomes(grange.counts)\n",
    "    atac_counts <- atac_counts[as.vector(grange.use), ]\n",
    "    annotations <- GetGRangesFromEnsDb(ensdb = EnsDb.Hsapiens.v86)\n",
    "    seqlevelsStyle(annotations) <- 'UCSC'\n",
    "    genome(annotations) <- \"hg38\"\n",
    "\n",
    "    chrom_assay <- CreateChromatinAssay(\n",
    "       counts = atac_counts,\n",
    "       sep = c(\":\", \"-\"),\n",
    "       genome = 'hg38',\n",
    "       fragments = sampleFile_atac,\n",
    "       min.cells = 10,\n",
    "       annotation = annotations\n",
    "     )\n",
    "    seurat_filtered[[\"ATAC\"]] <- chrom_assay\n",
    "    \n",
    "    print(paste0('After doublet filtering: ', sampleName))\n",
    "    print(seurat_filtered)\n",
    "\n",
    "    filtered_list[[sampleName]] <- seurat_filtered\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "46f4612e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Reprocessing: AF_1_LA\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric\n",
      "To use Python UMAP via reticulate, set umap.method to 'umap-learn' and metric to 'correlation'\n",
      "This message will be shown once per session\"\n",
      "20:22:49 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "20:22:49 Read 4576 rows and found 50 numeric columns\n",
      "\n",
      "20:22:49 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "20:22:49 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "20:22:49 Writing NN index file to temp file /tmp/RtmpET4Vi4/file7d03c141351\n",
      "\n",
      "20:22:49 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "20:22:51 Annoy recall = 100%\n",
      "\n",
      "20:22:55 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "20:23:00 Found 2 connected components, \n",
      "falling back to 'spca' initialization with init_sdev = 1\n",
      "\n",
      "20:23:00 Using 'irlba' for PCA\n",
      "\n",
      "20:23:00 PCA: 2 components explained 37.72% variance\n",
      "\n",
      "20:23:00 Scaling init to sdev = 1\n",
      "\n",
      "20:23:01 Commencing optimization for 500 epochs, with 199516 positive edges\n",
      "\n",
      "20:23:09 Optimization finished\n",
      "\n",
      "Performing TF-IDF normalization\n",
      "\n",
      "Running SVD\n",
      "\n",
      "Scaling cell embeddings\n",
      "\n",
      "20:23:39 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "20:23:39 Read 4576 rows and found 49 numeric columns\n",
      "\n",
      "20:23:39 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "20:23:39 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "20:23:40 Writing NN index file to temp file /tmp/RtmpET4Vi4/file7d047021232\n",
      "\n",
      "20:23:40 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "20:23:41 Annoy recall = 100%\n",
      "\n",
      "20:23:45 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "20:23:51 Initializing from normalized Laplacian + noise (using irlba)\n",
      "\n",
      "20:23:51 Commencing optimization for 500 epochs, with 199472 positive edges\n",
      "\n",
      "20:23:59 Optimization finished\n",
      "\n",
      "Calculating cell-specific modality weights\n",
      "\n",
      "Finding 20 nearest neighbors for each modality.\n",
      "\n",
      "Calculating kernel bandwidths\n",
      "\n",
      "Finding multimodal neighbors\n",
      "\n",
      "Constructing multimodal KNN graph\n",
      "\n",
      "Constructing multimodal SNN graph\n",
      "\n",
      "20:24:13 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "20:24:16 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 20\n",
      "\n",
      "20:24:22 Initializing from normalized Laplacian + noise (using irlba)\n",
      "\n",
      "20:24:22 Commencing optimization for 500 epochs, with 143888 positive edges\n",
      "\n",
      "20:24:31 Optimization finished\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Reprocessing: AF_2_LA\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "20:25:25 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "20:25:25 Read 2471 rows and found 50 numeric columns\n",
      "\n",
      "20:25:25 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "20:25:25 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "20:25:25 Writing NN index file to temp file /tmp/RtmpET4Vi4/file7d0729934ed\n",
      "\n",
      "20:25:25 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "20:25:26 Annoy recall = 100%\n",
      "\n",
      "20:25:30 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "20:25:35 Initializing from normalized Laplacian + noise (using irlba)\n",
      "\n",
      "20:25:36 Commencing optimization for 500 epochs, with 103670 positive edges\n",
      "\n",
      "20:25:42 Optimization finished\n",
      "\n",
      "Performing TF-IDF normalization\n",
      "\n",
      "Running SVD\n",
      "\n",
      "Scaling cell embeddings\n",
      "\n",
      "20:25:54 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "20:25:54 Read 2471 rows and found 49 numeric columns\n",
      "\n",
      "20:25:54 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "20:25:54 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "20:25:55 Writing NN index file to temp file /tmp/RtmpET4Vi4/file7d05022e774\n",
      "\n",
      "20:25:55 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "20:25:55 Annoy recall = 100%\n",
      "\n",
      "20:25:59 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "20:26:05 Initializing from normalized Laplacian + noise (using irlba)\n",
      "\n",
      "20:26:05 Commencing optimization for 500 epochs, with 100954 positive edges\n",
      "\n",
      "20:26:10 Optimization finished\n",
      "\n",
      "Calculating cell-specific modality weights\n",
      "\n",
      "Finding 20 nearest neighbors for each modality.\n",
      "\n",
      "Calculating kernel bandwidths\n",
      "\n",
      "Finding multimodal neighbors\n",
      "\n",
      "Constructing multimodal KNN graph\n",
      "\n",
      "Constructing multimodal SNN graph\n",
      "\n",
      "20:26:17 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "20:26:20 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 20\n",
      "\n",
      "20:26:25 Initializing from normalized Laplacian + noise (using irlba)\n",
      "\n",
      "20:26:25 Commencing optimization for 500 epochs, with 74088 positive edges\n",
      "\n",
      "20:26:31 Optimization finished\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Reprocessing: AF_3_LA\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "20:27:05 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "20:27:05 Read 7640 rows and found 50 numeric columns\n",
      "\n",
      "20:27:05 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "20:27:05 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "20:27:07 Writing NN index file to temp file /tmp/RtmpET4Vi4/file7d046adfba4\n",
      "\n",
      "20:27:07 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "20:27:09 Annoy recall = 100%\n",
      "\n",
      "20:27:13 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "20:27:20 Initializing from normalized Laplacian + noise (using irlba)\n",
      "\n",
      "20:27:21 Commencing optimization for 500 epochs, with 344834 positive edges\n",
      "\n",
      "20:27:34 Optimization finished\n",
      "\n",
      "Performing TF-IDF normalization\n",
      "\n",
      "Running SVD\n",
      "\n",
      "Scaling cell embeddings\n",
      "\n",
      "20:28:17 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "20:28:17 Read 7640 rows and found 49 numeric columns\n",
      "\n",
      "20:28:17 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "20:28:17 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "20:28:19 Writing NN index file to temp file /tmp/RtmpET4Vi4/file7d02e16954e\n",
      "\n",
      "20:28:19 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "20:28:22 Annoy recall = 100%\n",
      "\n",
      "20:28:26 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "20:28:32 Initializing from normalized Laplacian + noise (using irlba)\n",
      "\n",
      "20:28:32 Commencing optimization for 500 epochs, with 359348 positive edges\n",
      "\n",
      "20:28:44 Optimization finished\n",
      "\n",
      "Calculating cell-specific modality weights\n",
      "\n",
      "Finding 20 nearest neighbors for each modality.\n",
      "\n",
      "Calculating kernel bandwidths\n",
      "\n",
      "Finding multimodal neighbors\n",
      "\n",
      "Constructing multimodal KNN graph\n",
      "\n",
      "Constructing multimodal SNN graph\n",
      "\n",
      "20:29:07 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "20:29:10 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 20\n",
      "\n",
      "20:29:15 Initializing from normalized Laplacian + noise (using irlba)\n",
      "\n",
      "20:29:16 Commencing optimization for 500 epochs, with 250630 positive edges\n",
      "\n",
      "20:29:30 Optimization finished\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Reprocessing: AF_4_LA\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "20:30:40 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "20:30:40 Read 6124 rows and found 50 numeric columns\n",
      "\n",
      "20:30:40 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "20:30:40 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "20:30:41 Writing NN index file to temp file /tmp/RtmpET4Vi4/file7d044d5ef28\n",
      "\n",
      "20:30:41 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "20:30:43 Annoy recall = 100%\n",
      "\n",
      "20:30:48 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "20:30:53 Initializing from normalized Laplacian + noise (using irlba)\n",
      "\n",
      "20:30:54 Commencing optimization for 500 epochs, with 262112 positive edges\n",
      "\n",
      "20:31:05 Optimization finished\n",
      "\n",
      "Performing TF-IDF normalization\n",
      "\n",
      "Running SVD\n",
      "\n",
      "Scaling cell embeddings\n",
      "\n",
      "20:31:50 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "20:31:50 Read 6124 rows and found 49 numeric columns\n",
      "\n",
      "20:31:50 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "20:31:50 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "20:31:52 Writing NN index file to temp file /tmp/RtmpET4Vi4/file7d06e30bd72\n",
      "\n",
      "20:31:52 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "20:31:54 Annoy recall = 100%\n",
      "\n",
      "20:31:58 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "20:32:04 Initializing from normalized Laplacian + noise (using irlba)\n",
      "\n",
      "20:32:04 Commencing optimization for 500 epochs, with 260096 positive edges\n",
      "\n",
      "20:32:15 Optimization finished\n",
      "\n",
      "Calculating cell-specific modality weights\n",
      "\n",
      "Finding 20 nearest neighbors for each modality.\n",
      "\n",
      "Calculating kernel bandwidths\n",
      "\n",
      "Finding multimodal neighbors\n",
      "\n",
      "Constructing multimodal KNN graph\n",
      "\n",
      "Constructing multimodal SNN graph\n",
      "\n",
      "20:32:34 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "20:32:36 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 20\n",
      "\n",
      "20:32:42 Found 2 connected components, \n",
      "falling back to random initialization\n",
      "\n",
      "20:32:42 Initializing from uniform random\n",
      "\n",
      "20:32:42 Commencing optimization for 500 epochs, with 191058 positive edges\n",
      "\n",
      "20:32:53 Optimization finished\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Reprocessing: AF_5_LA\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "20:34:01 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "20:34:01 Read 12580 rows and found 50 numeric columns\n",
      "\n",
      "20:34:01 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "20:34:01 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "20:34:04 Writing NN index file to temp file /tmp/RtmpET4Vi4/file7d06fd3d577\n",
      "\n",
      "20:34:04 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "20:34:08 Annoy recall = 100%\n",
      "\n",
      "20:34:12 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "20:34:18 Initializing from normalized Laplacian + noise (using irlba)\n",
      "\n",
      "20:34:19 Commencing optimization for 200 epochs, with 561006 positive edges\n",
      "\n",
      "20:34:27 Optimization finished\n",
      "\n",
      "Performing TF-IDF normalization\n",
      "\n",
      "Running SVD\n",
      "\n",
      "Scaling cell embeddings\n",
      "\n",
      "20:36:12 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "20:36:12 Read 12580 rows and found 49 numeric columns\n",
      "\n",
      "20:36:12 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "20:36:12 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "20:36:14 Writing NN index file to temp file /tmp/RtmpET4Vi4/file7d0230bf78b\n",
      "\n",
      "20:36:14 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "20:36:20 Annoy recall = 100%\n",
      "\n",
      "20:36:24 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "20:36:30 Initializing from normalized Laplacian + noise (using irlba)\n",
      "\n",
      "20:36:30 Commencing optimization for 200 epochs, with 540392 positive edges\n",
      "\n",
      "20:36:39 Optimization finished\n",
      "\n",
      "Calculating cell-specific modality weights\n",
      "\n",
      "Finding 20 nearest neighbors for each modality.\n",
      "\n",
      "Calculating kernel bandwidths\n",
      "\n",
      "Finding multimodal neighbors\n",
      "\n",
      "Constructing multimodal KNN graph\n",
      "\n",
      "Constructing multimodal SNN graph\n",
      "\n",
      "20:37:20 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "20:37:24 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 20\n",
      "\n",
      "20:37:30 Initializing from normalized Laplacian + noise (using irlba)\n",
      "\n",
      "20:37:30 Commencing optimization for 200 epochs, with 402482 positive edges\n",
      "\n",
      "20:37:40 Optimization finished\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Reprocessing: AF_6_LA\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "20:39:39 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "20:39:39 Read 9728 rows and found 50 numeric columns\n",
      "\n",
      "20:39:39 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "20:39:39 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "20:39:41 Writing NN index file to temp file /tmp/RtmpET4Vi4/file7d07908d92f\n",
      "\n",
      "20:39:41 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "20:39:45 Annoy recall = 100%\n",
      "\n",
      "20:39:49 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "20:39:55 Initializing from normalized Laplacian + noise (using irlba)\n",
      "\n",
      "20:39:56 Commencing optimization for 500 epochs, with 423854 positive edges\n",
      "\n",
      "20:40:10 Optimization finished\n",
      "\n",
      "Performing TF-IDF normalization\n",
      "\n",
      "Running SVD\n",
      "\n",
      "Scaling cell embeddings\n",
      "\n",
      "20:41:16 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "20:41:16 Read 9728 rows and found 49 numeric columns\n",
      "\n",
      "20:41:16 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "20:41:16 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "20:41:17 Writing NN index file to temp file /tmp/RtmpET4Vi4/file7d0445a920\n",
      "\n",
      "20:41:17 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "20:41:21 Annoy recall = 100%\n",
      "\n",
      "20:41:25 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "20:41:31 Initializing from normalized Laplacian + noise (using irlba)\n",
      "\n",
      "20:41:32 Commencing optimization for 500 epochs, with 425368 positive edges\n",
      "\n",
      "20:41:46 Optimization finished\n",
      "\n",
      "Calculating cell-specific modality weights\n",
      "\n",
      "Finding 20 nearest neighbors for each modality.\n",
      "\n",
      "Calculating kernel bandwidths\n",
      "\n",
      "Finding multimodal neighbors\n",
      "\n",
      "Constructing multimodal KNN graph\n",
      "\n",
      "Constructing multimodal SNN graph\n",
      "\n",
      "20:42:17 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "20:42:21 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 20\n",
      "\n",
      "20:42:26 Initializing from normalized Laplacian + noise (using irlba)\n",
      "\n",
      "20:42:27 Commencing optimization for 500 epochs, with 305778 positive edges\n",
      "\n",
      "20:42:43 Optimization finished\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Reprocessing: AF_7_LA\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "20:44:23 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "20:44:23 Read 8582 rows and found 50 numeric columns\n",
      "\n",
      "20:44:23 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "20:44:23 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "20:44:24 Writing NN index file to temp file /tmp/RtmpET4Vi4/file7d0137f2a4\n",
      "\n",
      "20:44:24 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "20:44:27 Annoy recall = 100%\n",
      "\n",
      "20:44:31 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "20:44:37 Initializing from normalized Laplacian + noise (using irlba)\n",
      "\n",
      "20:44:40 Commencing optimization for 500 epochs, with 383468 positive edges\n",
      "\n",
      "20:44:53 Optimization finished\n",
      "\n",
      "Performing TF-IDF normalization\n",
      "\n",
      "Running SVD\n",
      "\n",
      "Scaling cell embeddings\n",
      "\n",
      "20:45:50 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "20:45:50 Read 8582 rows and found 49 numeric columns\n",
      "\n",
      "20:45:50 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "20:45:50 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "20:45:51 Writing NN index file to temp file /tmp/RtmpET4Vi4/file7d016174ac\n",
      "\n",
      "20:45:51 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "20:45:55 Annoy recall = 100%\n",
      "\n",
      "20:45:59 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "20:46:05 Initializing from normalized Laplacian + noise (using irlba)\n",
      "\n",
      "20:46:05 Commencing optimization for 500 epochs, with 384428 positive edges\n",
      "\n",
      "20:46:18 Optimization finished\n",
      "\n",
      "Calculating cell-specific modality weights\n",
      "\n",
      "Finding 20 nearest neighbors for each modality.\n",
      "\n",
      "Calculating kernel bandwidths\n",
      "\n",
      "Finding multimodal neighbors\n",
      "\n",
      "Constructing multimodal KNN graph\n",
      "\n",
      "Constructing multimodal SNN graph\n",
      "\n",
      "20:46:44 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "20:46:48 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 20\n",
      "\n",
      "20:46:53 Initializing from normalized Laplacian + noise (using irlba)\n",
      "\n",
      "20:46:54 Commencing optimization for 500 epochs, with 276298 positive edges\n",
      "\n",
      "20:47:09 Optimization finished\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Reprocessing: AF_8_LA\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "20:48:30 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "20:48:30 Read 7281 rows and found 50 numeric columns\n",
      "\n",
      "20:48:30 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "20:48:30 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "20:48:31 Writing NN index file to temp file /tmp/RtmpET4Vi4/file7d04456da19\n",
      "\n",
      "20:48:31 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "20:48:34 Annoy recall = 100%\n",
      "\n",
      "20:48:38 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "20:48:43 Initializing from normalized Laplacian + noise (using irlba)\n",
      "\n",
      "20:48:47 Commencing optimization for 500 epochs, with 318362 positive edges\n",
      "\n",
      "20:48:58 Optimization finished\n",
      "\n",
      "Performing TF-IDF normalization\n",
      "\n",
      "Running SVD\n",
      "\n",
      "Scaling cell embeddings\n",
      "\n",
      "20:50:14 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "20:50:14 Read 7281 rows and found 49 numeric columns\n",
      "\n",
      "20:50:14 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "20:50:14 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "20:50:15 Writing NN index file to temp file /tmp/RtmpET4Vi4/file7d0293e6f32\n",
      "\n",
      "20:50:15 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "20:50:17 Annoy recall = 100%\n",
      "\n",
      "20:50:22 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "20:50:28 Initializing from normalized Laplacian + noise (using irlba)\n",
      "\n",
      "20:50:28 Commencing optimization for 500 epochs, with 326026 positive edges\n",
      "\n",
      "20:50:40 Optimization finished\n",
      "\n",
      "Calculating cell-specific modality weights\n",
      "\n",
      "Finding 20 nearest neighbors for each modality.\n",
      "\n",
      "Calculating kernel bandwidths\n",
      "\n",
      "Finding multimodal neighbors\n",
      "\n",
      "Constructing multimodal KNN graph\n",
      "\n",
      "Constructing multimodal SNN graph\n",
      "\n",
      "20:51:02 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "20:51:05 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 20\n",
      "\n",
      "20:51:11 Initializing from normalized Laplacian + noise (using irlba)\n",
      "\n",
      "20:51:12 Commencing optimization for 500 epochs, with 228732 positive edges\n",
      "\n",
      "20:51:24 Optimization finished\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Reprocessing: CTRL_1_LA\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "20:53:03 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "20:53:03 Read 5881 rows and found 50 numeric columns\n",
      "\n",
      "20:53:03 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "20:53:03 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "20:53:04 Writing NN index file to temp file /tmp/RtmpET4Vi4/file7d054837a11\n",
      "\n",
      "20:53:04 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "20:53:06 Annoy recall = 100%\n",
      "\n",
      "20:53:10 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "20:53:16 Initializing from normalized Laplacian + noise (using irlba)\n",
      "\n",
      "20:53:17 Commencing optimization for 500 epochs, with 254986 positive edges\n",
      "\n",
      "20:53:27 Optimization finished\n",
      "\n",
      "Performing TF-IDF normalization\n",
      "\n",
      "Running SVD\n",
      "\n",
      "Scaling cell embeddings\n",
      "\n",
      "20:54:17 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "20:54:17 Read 5881 rows and found 49 numeric columns\n",
      "\n",
      "20:54:17 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "20:54:17 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "20:54:18 Writing NN index file to temp file /tmp/RtmpET4Vi4/file7d04c1081d2\n",
      "\n",
      "20:54:18 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "20:54:20 Annoy recall = 100%\n",
      "\n",
      "20:54:24 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "20:54:30 Initializing from normalized Laplacian + noise (using irlba)\n",
      "\n",
      "20:54:30 Commencing optimization for 500 epochs, with 259034 positive edges\n",
      "\n",
      "20:54:40 Optimization finished\n",
      "\n",
      "Calculating cell-specific modality weights\n",
      "\n",
      "Finding 20 nearest neighbors for each modality.\n",
      "\n",
      "Calculating kernel bandwidths\n",
      "\n",
      "Finding multimodal neighbors\n",
      "\n",
      "Constructing multimodal KNN graph\n",
      "\n",
      "Constructing multimodal SNN graph\n",
      "\n",
      "20:54:58 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "20:55:01 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 20\n",
      "\n",
      "20:55:06 Initializing from normalized Laplacian + noise (using irlba)\n",
      "\n",
      "20:55:07 Commencing optimization for 500 epochs, with 186236 positive edges\n",
      "\n",
      "20:55:18 Optimization finished\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Reprocessing: CTRL_2_LA\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "20:56:30 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "20:56:30 Read 7433 rows and found 50 numeric columns\n",
      "\n",
      "20:56:30 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "20:56:30 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "20:56:31 Writing NN index file to temp file /tmp/RtmpET4Vi4/file7d064565085\n",
      "\n",
      "20:56:31 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "20:56:34 Annoy recall = 100%\n",
      "\n",
      "20:56:38 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "20:56:43 Found 2 connected components, \n",
      "falling back to 'spca' initialization with init_sdev = 1\n",
      "\n",
      "20:56:43 Using 'irlba' for PCA\n",
      "\n",
      "20:56:43 PCA: 2 components explained 44.62% variance\n",
      "\n",
      "20:56:43 Scaling init to sdev = 1\n",
      "\n",
      "20:56:43 Commencing optimization for 500 epochs, with 325820 positive edges\n",
      "\n",
      "20:56:55 Optimization finished\n",
      "\n",
      "Performing TF-IDF normalization\n",
      "\n",
      "Running SVD\n",
      "\n",
      "Scaling cell embeddings\n",
      "\n",
      "20:57:53 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "20:57:53 Read 7433 rows and found 49 numeric columns\n",
      "\n",
      "20:57:53 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "20:57:53 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "20:57:54 Writing NN index file to temp file /tmp/RtmpET4Vi4/file7d012c424f0\n",
      "\n",
      "20:57:54 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "20:57:57 Annoy recall = 100%\n",
      "\n",
      "20:58:01 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "20:58:07 Initializing from normalized Laplacian + noise (using irlba)\n",
      "\n",
      "20:58:07 Commencing optimization for 500 epochs, with 334850 positive edges\n",
      "\n",
      "20:58:19 Optimization finished\n",
      "\n",
      "Calculating cell-specific modality weights\n",
      "\n",
      "Finding 20 nearest neighbors for each modality.\n",
      "\n",
      "Calculating kernel bandwidths\n",
      "\n",
      "Finding multimodal neighbors\n",
      "\n",
      "Constructing multimodal KNN graph\n",
      "\n",
      "Constructing multimodal SNN graph\n",
      "\n",
      "20:58:41 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "20:58:44 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 20\n",
      "\n",
      "20:58:50 Found 2 connected components, \n",
      "falling back to random initialization\n",
      "\n",
      "20:58:50 Initializing from uniform random\n",
      "\n",
      "20:58:50 Commencing optimization for 500 epochs, with 233830 positive edges\n",
      "\n",
      "20:59:02 Optimization finished\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Reprocessing: CTRL_3_LA\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "21:00:27 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "21:00:27 Read 8938 rows and found 50 numeric columns\n",
      "\n",
      "21:00:27 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "21:00:27 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "21:00:29 Writing NN index file to temp file /tmp/RtmpET4Vi4/file7d05e1cba0f\n",
      "\n",
      "21:00:29 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "21:00:32 Annoy recall = 100%\n",
      "\n",
      "21:00:36 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "21:00:42 Initializing from normalized Laplacian + noise (using irlba)\n",
      "\n",
      "21:00:43 Commencing optimization for 500 epochs, with 400914 positive edges\n",
      "\n",
      "21:00:56 Optimization finished\n",
      "\n",
      "Performing TF-IDF normalization\n",
      "\n",
      "Running SVD\n",
      "\n",
      "Scaling cell embeddings\n",
      "\n",
      "21:02:11 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "21:02:11 Read 8938 rows and found 49 numeric columns\n",
      "\n",
      "21:02:11 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "21:02:11 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "21:02:13 Writing NN index file to temp file /tmp/RtmpET4Vi4/file7d0604b85ed\n",
      "\n",
      "21:02:13 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "21:02:16 Annoy recall = 100%\n",
      "\n",
      "21:02:20 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "21:02:26 Initializing from normalized Laplacian + noise (using irlba)\n",
      "\n",
      "21:02:26 Commencing optimization for 500 epochs, with 419704 positive edges\n",
      "\n",
      "21:02:40 Optimization finished\n",
      "\n",
      "Calculating cell-specific modality weights\n",
      "\n",
      "Finding 20 nearest neighbors for each modality.\n",
      "\n",
      "Calculating kernel bandwidths\n",
      "\n",
      "Finding multimodal neighbors\n",
      "\n",
      "Constructing multimodal KNN graph\n",
      "\n",
      "Constructing multimodal SNN graph\n",
      "\n",
      "21:03:07 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "21:03:10 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 20\n",
      "\n",
      "21:03:16 Initializing from normalized Laplacian + noise (using irlba)\n",
      "\n",
      "21:03:17 Commencing optimization for 500 epochs, with 290580 positive edges\n",
      "\n",
      "21:03:31 Optimization finished\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Reprocessing: CTRL_4_LA\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "21:05:09 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "21:05:09 Read 7794 rows and found 50 numeric columns\n",
      "\n",
      "21:05:09 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "21:05:09 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "21:05:10 Writing NN index file to temp file /tmp/RtmpET4Vi4/file7d018cb3da6\n",
      "\n",
      "21:05:10 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "21:05:13 Annoy recall = 100%\n",
      "\n",
      "21:05:17 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "21:05:22 Initializing from normalized Laplacian + noise (using irlba)\n",
      "\n",
      "21:05:23 Commencing optimization for 500 epochs, with 348850 positive edges\n",
      "\n",
      "21:05:35 Optimization finished\n",
      "\n",
      "Performing TF-IDF normalization\n",
      "\n",
      "Running SVD\n",
      "\n",
      "Scaling cell embeddings\n",
      "\n",
      "21:06:19 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "21:06:19 Read 7794 rows and found 49 numeric columns\n",
      "\n",
      "21:06:19 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "21:06:19 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "21:06:20 Writing NN index file to temp file /tmp/RtmpET4Vi4/file7d05670f7d0\n",
      "\n",
      "21:06:20 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "21:06:23 Annoy recall = 100%\n",
      "\n",
      "21:06:27 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "21:06:33 Initializing from normalized Laplacian + noise (using irlba)\n",
      "\n",
      "21:06:33 Commencing optimization for 500 epochs, with 348968 positive edges\n",
      "\n",
      "21:06:45 Optimization finished\n",
      "\n",
      "Calculating cell-specific modality weights\n",
      "\n",
      "Finding 20 nearest neighbors for each modality.\n",
      "\n",
      "Calculating kernel bandwidths\n",
      "\n",
      "Finding multimodal neighbors\n",
      "\n",
      "Constructing multimodal KNN graph\n",
      "\n",
      "Constructing multimodal SNN graph\n",
      "\n",
      "21:07:10 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "21:07:13 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 20\n",
      "\n",
      "21:07:19 Initializing from normalized Laplacian + noise (using irlba)\n",
      "\n",
      "21:07:19 Commencing optimization for 500 epochs, with 250304 positive edges\n",
      "\n",
      "21:07:33 Optimization finished\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Reprocessing: CTRL_5_LA\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "21:08:47 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "21:08:47 Read 7152 rows and found 50 numeric columns\n",
      "\n",
      "21:08:47 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "21:08:47 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "21:08:49 Writing NN index file to temp file /tmp/RtmpET4Vi4/file7d036fe143b\n",
      "\n",
      "21:08:49 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "21:08:51 Annoy recall = 100%\n",
      "\n",
      "21:08:55 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "21:09:01 Initializing from normalized Laplacian + noise (using irlba)\n",
      "\n",
      "21:09:02 Commencing optimization for 500 epochs, with 302414 positive edges\n",
      "\n",
      "21:09:14 Optimization finished\n",
      "\n",
      "Performing TF-IDF normalization\n",
      "\n",
      "Running SVD\n",
      "\n",
      "Scaling cell embeddings\n",
      "\n",
      "21:09:58 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "21:09:58 Read 7152 rows and found 49 numeric columns\n",
      "\n",
      "21:09:58 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "21:09:58 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "21:09:59 Writing NN index file to temp file /tmp/RtmpET4Vi4/file7d03e710be7\n",
      "\n",
      "21:09:59 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "21:10:02 Annoy recall = 100%\n",
      "\n",
      "21:10:06 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "21:10:12 Initializing from normalized Laplacian + noise (using irlba)\n",
      "\n",
      "21:10:12 Commencing optimization for 500 epochs, with 323734 positive edges\n",
      "\n",
      "21:10:24 Optimization finished\n",
      "\n",
      "Calculating cell-specific modality weights\n",
      "\n",
      "Finding 20 nearest neighbors for each modality.\n",
      "\n",
      "Calculating kernel bandwidths\n",
      "\n",
      "Finding multimodal neighbors\n",
      "\n",
      "Constructing multimodal KNN graph\n",
      "\n",
      "Constructing multimodal SNN graph\n",
      "\n",
      "21:10:46 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "21:10:49 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 20\n",
      "\n",
      "21:10:55 Found 2 connected components, \n",
      "falling back to random initialization\n",
      "\n",
      "21:10:55 Initializing from uniform random\n",
      "\n",
      "21:10:55 Commencing optimization for 500 epochs, with 219772 positive edges\n",
      "\n",
      "21:11:07 Optimization finished\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Reprocessing: CTRL_6_LA\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "21:12:24 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "21:12:24 Read 9763 rows and found 50 numeric columns\n",
      "\n",
      "21:12:24 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "21:12:24 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "21:12:26 Writing NN index file to temp file /tmp/RtmpET4Vi4/file7d062c98f80\n",
      "\n",
      "21:12:26 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "21:12:29 Annoy recall = 100%\n",
      "\n",
      "21:12:33 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "21:12:39 Initializing from normalized Laplacian + noise (using irlba)\n",
      "\n",
      "21:12:40 Commencing optimization for 500 epochs, with 439046 positive edges\n",
      "\n",
      "21:12:55 Optimization finished\n",
      "\n",
      "Performing TF-IDF normalization\n",
      "\n",
      "Running SVD\n",
      "\n",
      "Scaling cell embeddings\n",
      "\n",
      "21:14:05 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "21:14:05 Read 9763 rows and found 49 numeric columns\n",
      "\n",
      "21:14:05 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "21:14:05 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "21:14:07 Writing NN index file to temp file /tmp/RtmpET4Vi4/file7d0182e0960\n",
      "\n",
      "21:14:07 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "21:14:11 Annoy recall = 100%\n",
      "\n",
      "21:14:15 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "21:14:21 Initializing from normalized Laplacian + noise (using irlba)\n",
      "\n",
      "21:14:21 Commencing optimization for 500 epochs, with 426008 positive edges\n",
      "\n",
      "21:14:36 Optimization finished\n",
      "\n",
      "Calculating cell-specific modality weights\n",
      "\n",
      "Finding 20 nearest neighbors for each modality.\n",
      "\n",
      "Calculating kernel bandwidths\n",
      "\n",
      "Finding multimodal neighbors\n",
      "\n",
      "Constructing multimodal KNN graph\n",
      "\n",
      "Constructing multimodal SNN graph\n",
      "\n",
      "21:15:07 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "21:15:10 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 20\n",
      "\n",
      "21:15:16 Initializing from normalized Laplacian + noise (using irlba)\n",
      "\n",
      "21:15:17 Commencing optimization for 500 epochs, with 310080 positive edges\n",
      "\n",
      "21:15:33 Optimization finished\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Reprocessing: CTRL_7_LA\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "21:17:15 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "21:17:15 Read 7152 rows and found 50 numeric columns\n",
      "\n",
      "21:17:15 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "21:17:15 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "21:17:16 Writing NN index file to temp file /tmp/RtmpET4Vi4/file7d03a764d0d\n",
      "\n",
      "21:17:16 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "21:17:18 Annoy recall = 100%\n",
      "\n",
      "21:17:22 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "21:17:28 Initializing from normalized Laplacian + noise (using irlba)\n",
      "\n",
      "21:17:30 Commencing optimization for 500 epochs, with 310008 positive edges\n",
      "\n",
      "21:17:42 Optimization finished\n",
      "\n",
      "Performing TF-IDF normalization\n",
      "\n",
      "Running SVD\n",
      "\n",
      "Scaling cell embeddings\n",
      "\n",
      "21:19:03 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "21:19:03 Read 7152 rows and found 49 numeric columns\n",
      "\n",
      "21:19:03 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "21:19:03 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "21:19:04 Writing NN index file to temp file /tmp/RtmpET4Vi4/file7d07a02d457\n",
      "\n",
      "21:19:04 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "21:19:07 Annoy recall = 100%\n",
      "\n",
      "21:19:11 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "21:19:16 Initializing from normalized Laplacian + noise (using irlba)\n",
      "\n",
      "21:19:17 Commencing optimization for 500 epochs, with 324022 positive edges\n",
      "\n",
      "21:19:28 Optimization finished\n",
      "\n",
      "Calculating cell-specific modality weights\n",
      "\n",
      "Finding 20 nearest neighbors for each modality.\n",
      "\n",
      "Calculating kernel bandwidths\n",
      "\n",
      "Finding multimodal neighbors\n",
      "\n",
      "Constructing multimodal KNN graph\n",
      "\n",
      "Constructing multimodal SNN graph\n",
      "\n",
      "21:19:50 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "21:19:53 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 20\n",
      "\n",
      "21:19:59 Found 2 connected components, \n",
      "falling back to random initialization\n",
      "\n",
      "21:19:59 Initializing from uniform random\n",
      "\n",
      "21:19:59 Commencing optimization for 500 epochs, with 224126 positive edges\n",
      "\n",
      "21:20:11 Optimization finished\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Reprocessing: CTRL_8_LA\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "21:21:58 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "21:21:58 Read 6047 rows and found 50 numeric columns\n",
      "\n",
      "21:21:58 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "21:21:58 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "21:21:59 Writing NN index file to temp file /tmp/RtmpET4Vi4/file7d024dab353\n",
      "\n",
      "21:21:59 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "21:22:01 Annoy recall = 100%\n",
      "\n",
      "21:22:05 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "21:22:11 Initializing from normalized Laplacian + noise (using irlba)\n",
      "\n",
      "21:22:12 Commencing optimization for 500 epochs, with 257856 positive edges\n",
      "\n",
      "21:22:23 Optimization finished\n",
      "\n",
      "Performing TF-IDF normalization\n",
      "\n",
      "Running SVD\n",
      "\n",
      "Scaling cell embeddings\n",
      "\n",
      "21:22:52 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "21:22:52 Read 6047 rows and found 49 numeric columns\n",
      "\n",
      "21:22:52 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "21:22:52 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "21:22:53 Writing NN index file to temp file /tmp/RtmpET4Vi4/file7d04ddf8a65\n",
      "\n",
      "21:22:53 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "21:22:56 Annoy recall = 100%\n",
      "\n",
      "21:23:00 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "21:23:06 Initializing from normalized Laplacian + noise (using irlba)\n",
      "\n",
      "21:23:06 Commencing optimization for 500 epochs, with 260072 positive edges\n",
      "\n",
      "21:23:16 Optimization finished\n",
      "\n",
      "Calculating cell-specific modality weights\n",
      "\n",
      "Finding 20 nearest neighbors for each modality.\n",
      "\n",
      "Calculating kernel bandwidths\n",
      "\n",
      "Finding multimodal neighbors\n",
      "\n",
      "Constructing multimodal KNN graph\n",
      "\n",
      "Constructing multimodal SNN graph\n",
      "\n",
      "21:23:34 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "21:23:37 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 20\n",
      "\n",
      "21:23:37 1 smooth knn distance failures\n",
      "\n",
      "21:23:43 Initializing from normalized Laplacian + noise (using irlba)\n",
      "\n",
      "21:23:43 Commencing optimization for 500 epochs, with 185740 positive edges\n",
      "\n",
      "21:23:54 Optimization finished\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Reprocessing: D9_LA\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "21:24:52 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "21:24:53 Read 7756 rows and found 50 numeric columns\n",
      "\n",
      "21:24:53 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "21:24:53 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "21:24:54 Writing NN index file to temp file /tmp/RtmpET4Vi4/file7d06115fe48\n",
      "\n",
      "21:24:54 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "21:24:56 Annoy recall = 100%\n",
      "\n",
      "21:25:00 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "21:25:06 Found 2 connected components, \n",
      "falling back to 'spca' initialization with init_sdev = 1\n",
      "\n",
      "21:25:06 Using 'irlba' for PCA\n",
      "\n",
      "21:25:06 PCA: 2 components explained 41.99% variance\n",
      "\n",
      "21:25:06 Scaling init to sdev = 1\n",
      "\n",
      "21:25:06 Commencing optimization for 500 epochs, with 337240 positive edges\n",
      "\n",
      "21:25:18 Optimization finished\n",
      "\n",
      "Performing TF-IDF normalization\n",
      "\n",
      "Running SVD\n",
      "\n",
      "Scaling cell embeddings\n",
      "\n",
      "21:26:15 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "21:26:15 Read 7756 rows and found 49 numeric columns\n",
      "\n",
      "21:26:15 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "21:26:15 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "21:26:17 Writing NN index file to temp file /tmp/RtmpET4Vi4/file7d02413bcbd\n",
      "\n",
      "21:26:17 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "21:26:19 Annoy recall = 100%\n",
      "\n",
      "21:26:24 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "21:26:29 Initializing from normalized Laplacian + noise (using irlba)\n",
      "\n",
      "21:26:30 Commencing optimization for 500 epochs, with 355874 positive edges\n",
      "\n",
      "21:26:42 Optimization finished\n",
      "\n",
      "Calculating cell-specific modality weights\n",
      "\n",
      "Finding 20 nearest neighbors for each modality.\n",
      "\n",
      "Calculating kernel bandwidths\n",
      "\n",
      "Finding multimodal neighbors\n",
      "\n",
      "Constructing multimodal KNN graph\n",
      "\n",
      "Constructing multimodal SNN graph\n",
      "\n",
      "21:27:05 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "21:27:08 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 20\n",
      "\n",
      "21:27:14 Initializing from normalized Laplacian + noise (using irlba)\n",
      "\n",
      "21:27:15 Commencing optimization for 500 epochs, with 244836 positive edges\n",
      "\n",
      "21:27:28 Optimization finished\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Reprocessing: D1_LA\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "21:28:42 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "21:28:42 Read 3862 rows and found 50 numeric columns\n",
      "\n",
      "21:28:42 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "21:28:42 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "21:28:43 Writing NN index file to temp file /tmp/RtmpET4Vi4/file7d07cd8079d\n",
      "\n",
      "21:28:43 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "21:28:44 Annoy recall = 100%\n",
      "\n",
      "21:28:48 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "21:28:54 Initializing from normalized Laplacian + noise (using irlba)\n",
      "\n",
      "21:28:54 Commencing optimization for 500 epochs, with 166886 positive edges\n",
      "\n",
      "21:29:01 Optimization finished\n",
      "\n",
      "Performing TF-IDF normalization\n",
      "\n",
      "Running SVD\n",
      "\n",
      "Scaling cell embeddings\n",
      "\n",
      "21:29:39 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "21:29:39 Read 3862 rows and found 49 numeric columns\n",
      "\n",
      "21:29:39 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "21:29:39 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "21:29:40 Writing NN index file to temp file /tmp/RtmpET4Vi4/file7d02379e6d1\n",
      "\n",
      "21:29:40 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "21:29:41 Annoy recall = 100%\n",
      "\n",
      "21:29:45 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "21:29:51 Initializing from normalized Laplacian + noise (using irlba)\n",
      "\n",
      "21:29:51 Commencing optimization for 500 epochs, with 168924 positive edges\n",
      "\n",
      "21:29:59 Optimization finished\n",
      "\n",
      "Calculating cell-specific modality weights\n",
      "\n",
      "Finding 20 nearest neighbors for each modality.\n",
      "\n",
      "Calculating kernel bandwidths\n",
      "\n",
      "Finding multimodal neighbors\n",
      "\n",
      "Constructing multimodal KNN graph\n",
      "\n",
      "Constructing multimodal SNN graph\n",
      "\n",
      "21:30:09 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "21:30:12 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 20\n",
      "\n",
      "21:30:18 Found 2 connected components, \n",
      "falling back to random initialization\n",
      "\n",
      "21:30:18 Initializing from uniform random\n",
      "\n",
      "21:30:18 Commencing optimization for 500 epochs, with 120216 positive edges\n",
      "\n",
      "21:30:26 Optimization finished\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Reprocessing: HF6_LA\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "21:31:19 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "21:31:19 Read 8200 rows and found 50 numeric columns\n",
      "\n",
      "21:31:19 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "21:31:19 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "21:31:20 Writing NN index file to temp file /tmp/RtmpET4Vi4/file7d02b21ed1e\n",
      "\n",
      "21:31:20 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "21:31:23 Annoy recall = 100%\n",
      "\n",
      "21:31:27 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "21:31:33 Found 2 connected components, \n",
      "falling back to 'spca' initialization with init_sdev = 1\n",
      "\n",
      "21:31:33 Using 'irlba' for PCA\n",
      "\n",
      "21:31:33 PCA: 2 components explained 40.3% variance\n",
      "\n",
      "21:31:33 Scaling init to sdev = 1\n",
      "\n",
      "21:31:33 Commencing optimization for 500 epochs, with 361350 positive edges\n",
      "\n",
      "21:31:46 Optimization finished\n",
      "\n",
      "Performing TF-IDF normalization\n",
      "\n",
      "Running SVD\n",
      "\n",
      "Scaling cell embeddings\n",
      "\n",
      "21:32:32 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "21:32:32 Read 8200 rows and found 49 numeric columns\n",
      "\n",
      "21:32:32 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "21:32:32 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "21:32:34 Writing NN index file to temp file /tmp/RtmpET4Vi4/file7d013317e6\n",
      "\n",
      "21:32:34 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "21:32:37 Annoy recall = 100%\n",
      "\n",
      "21:32:41 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "21:32:47 Initializing from normalized Laplacian + noise (using irlba)\n",
      "\n",
      "21:32:47 Commencing optimization for 500 epochs, with 382748 positive edges\n",
      "\n",
      "21:33:00 Optimization finished\n",
      "\n",
      "Calculating cell-specific modality weights\n",
      "\n",
      "Finding 20 nearest neighbors for each modality.\n",
      "\n",
      "Calculating kernel bandwidths\n",
      "\n",
      "Finding multimodal neighbors\n",
      "\n",
      "Constructing multimodal KNN graph\n",
      "\n",
      "Constructing multimodal SNN graph\n",
      "\n",
      "21:33:25 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "21:33:29 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 20\n",
      "\n",
      "21:33:34 Found 2 connected components, \n",
      "falling back to random initialization\n",
      "\n",
      "21:33:34 Initializing from uniform random\n",
      "\n",
      "21:33:34 Commencing optimization for 500 epochs, with 265146 positive edges\n",
      "\n",
      "21:33:47 Optimization finished\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Reprocessing: HF9_LA\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "21:35:11 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "21:35:11 Read 6900 rows and found 50 numeric columns\n",
      "\n",
      "21:35:11 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "21:35:11 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "21:35:12 Writing NN index file to temp file /tmp/RtmpET4Vi4/file7d0573c950a\n",
      "\n",
      "21:35:12 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "21:35:15 Annoy recall = 100%\n",
      "\n",
      "21:35:19 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "21:35:25 Initializing from normalized Laplacian + noise (using irlba)\n",
      "\n",
      "21:35:27 Commencing optimization for 500 epochs, with 292042 positive edges\n",
      "\n",
      "21:35:38 Optimization finished\n",
      "\n",
      "Performing TF-IDF normalization\n",
      "\n",
      "Running SVD\n",
      "\n",
      "Scaling cell embeddings\n",
      "\n",
      "21:36:32 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "21:36:32 Read 6900 rows and found 49 numeric columns\n",
      "\n",
      "21:36:32 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "21:36:32 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "21:36:33 Writing NN index file to temp file /tmp/RtmpET4Vi4/file7d012995af8\n",
      "\n",
      "21:36:33 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "21:36:35 Annoy recall = 100%\n",
      "\n",
      "21:36:39 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "21:36:45 Initializing from normalized Laplacian + noise (using irlba)\n",
      "\n",
      "21:36:46 Commencing optimization for 500 epochs, with 301856 positive edges\n",
      "\n",
      "21:36:57 Optimization finished\n",
      "\n",
      "Calculating cell-specific modality weights\n",
      "\n",
      "Finding 20 nearest neighbors for each modality.\n",
      "\n",
      "Calculating kernel bandwidths\n",
      "\n",
      "Finding multimodal neighbors\n",
      "\n",
      "Constructing multimodal KNN graph\n",
      "\n",
      "Constructing multimodal SNN graph\n",
      "\n",
      "21:37:18 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "21:37:21 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 20\n",
      "\n",
      "21:37:26 Found 2 connected components, \n",
      "falling back to random initialization\n",
      "\n",
      "21:37:26 Initializing from uniform random\n",
      "\n",
      "21:37:26 Commencing optimization for 500 epochs, with 211158 positive edges\n",
      "\n",
      "21:37:39 Optimization finished\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# seurat pipeline on each object\n",
    "for (i in seq_along(filtered_list)){\n",
    "    sampleName <- names(filtered_list[i])\n",
    "    seurat <- filtered_list[[i]]\n",
    "    print(paste0('Reprocessing: ', sampleName))\n",
    "    \n",
    "    # RNA analysis\n",
    "    DefaultAssay(seurat) <- 'RNA'\n",
    "    seurat <- NormalizeData(seurat, verbose = FALSE)\n",
    "    seurat <- FindVariableFeatures(seurat, verbose = FALSE)\n",
    "    seurat <- ScaleData(seurat, verbose = FALSE)\n",
    "    seurat <- RunPCA(seurat, verbose = FALSE)\n",
    "    seurat <- RunUMAP(seurat, dims = 1:50, reduction.name = 'umap.rna', reduction.key = 'rnaUMAP_')\n",
    "    \n",
    "    # ATAC analysis\n",
    "    # We exclude the first dimension as this is typically correlated with sequencing depth\n",
    "    DefaultAssay(seurat) <- 'ATAC'\n",
    "    seurat <- RunTFIDF(seurat)\n",
    "    seurat <- FindTopFeatures(seurat, min.cutoff='q0')\n",
    "    seurat <- RunSVD(seurat)\n",
    "    seurat <- RunUMAP(seurat, reduction='lsi', dims=2:50, reduction.name='umap.atac', reduction.key='atacUMAP_')\n",
    "    \n",
    "    # WNN multimodal analysis, representing a weighted combination of RNA and ATAC-seq modalities. \n",
    "    seurat <- FindMultiModalNeighbors(seurat, reduction.list=list('pca', 'lsi'), dims.list=list(1:50, 2:50))\n",
    "    seurat <- RunUMAP(seurat, nn.name='weighted.nn', reduction.name='umap.wnn', reduction.key='wnnUMAP_')\n",
    "    seurat <- FindClusters(seurat, graph.name='wsnn', verbose=FALSE)\n",
    "    \n",
    "    pdf(paste0(out_dir,'UMAPs_',sampleName,'_postDoublet+RNAcontamFilt.pdf'), height=7, width=7)\n",
    "    print(DimPlot(seurat, reduction = \"umap.rna\", label=TRUE, repel=TRUE))\n",
    "    print(DimPlot(seurat, reduction = \"umap.atac\", label=TRUE, repel=TRUE))\n",
    "    print(DimPlot(seurat, reduction = \"umap.wnn\", label=TRUE, repel=TRUE))\n",
    "    dev.off()\n",
    "\n",
    "    filtered_list[[sampleName]] <- seurat\n",
    "    \n",
    "    saveRDS(seurat, paste0(out_dir,sampleName,'_postDoublet+RNAcontamFilt.rds'))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa5f012e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
